{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b83cfacd",
   "metadata": {},
   "source": [
    "# Compare 6m and 2y auspice haplotypes; design library\n",
    "Using the 6m and 2y Nextstrain trees, we identified all `derived haplotypes` and their representative strains.\n",
    "Now we will compare those haplotypes. \n",
    "We will keep the 6m representative strains and toss the 2y strains. \n",
    "\n",
    "Then, we will compare to Nextclade metadata.\n",
    "First I'll purge all haplotypes from the Nextclade set that haven't been sampled in 2023. \n",
    "Then I'll merge haplotype record count with the Nextstrain dataset. \n",
    "We will purge any 2y haplotypes that have no record from 2023.\n",
    "\n",
    "This particular year there were only **37 Nexstrain derived haplotypes** that matched the given sampling requirements. A `derived haplotype` is a Nextstrain-specific piece of jargon that is assigned to an HA1 haplotype (a subclade with additional amino acid mutations called on that background, for example, a J.2 virus with S145N mutaiton would be called 'J.2:S145N') that has achieved some threshold of child/descendent sequences over a specified timeline. \n",
    "\n",
    "To more fully take advantage of the space in our library, we decided to use the Nextclade dataset to add additional haplotypes that had not achieved this threshold, and therefore were not called as `derived haplotypes` in the Nextstrain dataset.\n",
    "We opted to simply make a list of the highest frequency haplotypes in the Nextclade dataset, then remove those haplotypes already sampled by our Nextstrain analysis (ie, remove all `derived haplotypes`).\n",
    "We chose representative strains for each of those haplotypes by selecting the most recent strain (ie, most recent collection date) per haplotype. \n",
    "This **Nextclade method identified 25 additional haplotypes**.\n",
    "\n",
    "Overall, we selected **62 total representative strains**.\n",
    "\n",
    "Then, we will finalize the list of haplotypes and representative strains. \n",
    "I will output this as a large dataframe and a list of GISAID accession numbers. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1a9d74c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "192be9b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "resultsdir = '../results'\n",
    "datadir = '../data'\n",
    "os.makedirs(resultsdir, exist_ok=True)\n",
    "os.makedirs(datadir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e635a7fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is some basic info on the Auspice/Phylogenetic method-selected strains...\n",
      "the number output (6m + 2y) haplotypes is... 99\n",
      "the number of unique haplotypes occurring is... 67\n",
      "the number of overlapping haplotypes is... 16\n"
     ]
    }
   ],
   "source": [
    "# [Phylogenetic method]\n",
    "# My inputs are 2 TSV outputs  because I ran on the 2y (two year) and 6m (six month) trees for H3N2\n",
    "files = {'two_year': 'representative_strains_per_haplotype_2y.tsv', \n",
    "        'six_month': 'representative_strains_per_haplotype_6m.tsv'}\n",
    "\n",
    "# Code for identifying clade overlap \n",
    "strains_df = pd.DataFrame()\n",
    "for key in files: \n",
    "    # Import emerging clades output\n",
    "    df = pd.read_csv(os.path.join(datadir, 'auspice_haplotypes', files[key]), sep='\\t')\n",
    "    df['method'] = key\n",
    "    # Select minimum branch length strains\n",
    "    strains_df = pd.concat([strains_df, df[['name', 'div', 'num_date',\n",
    "    'clade_membership', 'subclade', 'haplotype', 'method']]])\n",
    "\n",
    "print('Here is some basic info on the Auspice/Phylogenetic method-selected strains...')\n",
    "# All haploytpes\n",
    "haplotypes = strains_df['haplotype'].tolist()\n",
    "print(f'the number output (6m + 2y) haplotypes is... {len(haplotypes)}')\n",
    "# Unique haplotypes\n",
    "unique_haplotypes = [i for i in haplotypes if haplotypes.count(i)==1]\n",
    "print(f'the number of unique haplotypes occurring is... {len(unique_haplotypes)}')\n",
    "# Haplotypes that appear both in 2y and 6m trees\n",
    "overlapping_haplotypes = list(set(haplotypes).difference(unique_haplotypes))\n",
    "print(f'the number of overlapping haplotypes is... {len(overlapping_haplotypes)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "be4056bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the overlapping haplotypes:\n",
      "['G.1.1.2', 'G.1.1.2:326R', 'G.1.3.1.1', 'G.1.3.1.1:122D', 'G.1.3.1.1:122D,144N,276E', 'G.1.3.1.1:122D,276E', 'G.1.3.1.1:124N,145N', 'G.1.3.1.1:145N', 'G.1.3.1.1:173R,276E', 'G.1.3.1.1:21T', 'G.1.3.1.1:242M', 'G.1.3.1.1:25V', 'G.1.3.2', 'G.2', 'G.2.1', 'G.2:242M']\n",
      "\n",
      "Now we can compare the number of strains from 6-month and 2-year, pre and post filtering.\n",
      "We should have kept all 6 month strains, and only a subselection of 2-year, if there was overlap.\n",
      "\n",
      "Here is the input:\n",
      "initial 6 month strains =  31\n",
      "initial 2 year strains =  68\n",
      "\n",
      "After filtering:\n",
      "the 6 month strains remaining =  31\n",
      "the 2 year strains remaining =  52\n",
      "therefore, the number of unique [Phylogenetic method] strains =  83\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>div</th>\n",
       "      <th>num_date</th>\n",
       "      <th>clade_membership</th>\n",
       "      <th>subclade</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Finland/391/2023</td>\n",
       "      <td>0.033588</td>\n",
       "      <td>2023.513699</td>\n",
       "      <td>2a.1b</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>six_month</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Kanagawa/IC2239/2023</td>\n",
       "      <td>0.034155</td>\n",
       "      <td>2023.486301</td>\n",
       "      <td>2a.1b</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>G.1.1.2:326R</td>\n",
       "      <td>six_month</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A/Singapore/NUH0526/2023</td>\n",
       "      <td>0.038133</td>\n",
       "      <td>2023.450685</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1</td>\n",
       "      <td>G.1.3.1.1</td>\n",
       "      <td>six_month</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A/Sydney/332/2023</td>\n",
       "      <td>0.039833</td>\n",
       "      <td>2023.450685</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1</td>\n",
       "      <td>G.1.3.1.1:122D</td>\n",
       "      <td>six_month</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/Bangkok/P3755/2023</td>\n",
       "      <td>0.040966</td>\n",
       "      <td>2023.683562</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1</td>\n",
       "      <td>G.1.3.1.1:122D,144N,276E</td>\n",
       "      <td>six_month</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>A/Netherlands/1685/2023</td>\n",
       "      <td>0.040403</td>\n",
       "      <td>2023.686301</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1</td>\n",
       "      <td>G.1.3.1.1:5E</td>\n",
       "      <td>six_month</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>A/SouthAfrica/R06477/2023</td>\n",
       "      <td>0.039833</td>\n",
       "      <td>2023.417808</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1</td>\n",
       "      <td>G.1.3.1.1:5V</td>\n",
       "      <td>six_month</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>A/Oman/3011/2023</td>\n",
       "      <td>0.039266</td>\n",
       "      <td>2023.543836</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1</td>\n",
       "      <td>G.1.3.1.1:78D</td>\n",
       "      <td>six_month</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>A/California/81/2023</td>\n",
       "      <td>0.039266</td>\n",
       "      <td>2023.628767</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1</td>\n",
       "      <td>G.1.3.1.1:81D</td>\n",
       "      <td>six_month</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>A/SouthAfrica/R05384/2023</td>\n",
       "      <td>0.037561</td>\n",
       "      <td>2023.371233</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2</td>\n",
       "      <td>G.2:173L,212T,242M</td>\n",
       "      <td>six_month</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         name       div     num_date clade_membership  \\\n",
       "0          A/Finland/391/2023  0.033588  2023.513699            2a.1b   \n",
       "1      A/Kanagawa/IC2239/2023  0.034155  2023.486301            2a.1b   \n",
       "2    A/Singapore/NUH0526/2023  0.038133  2023.450685          2a.3a.1   \n",
       "3           A/Sydney/332/2023  0.039833  2023.450685          2a.3a.1   \n",
       "4        A/Bangkok/P3755/2023  0.040966  2023.683562          2a.3a.1   \n",
       "..                        ...       ...          ...              ...   \n",
       "78    A/Netherlands/1685/2023  0.040403  2023.686301          2a.3a.1   \n",
       "79  A/SouthAfrica/R06477/2023  0.039833  2023.417808          2a.3a.1   \n",
       "80           A/Oman/3011/2023  0.039266  2023.543836          2a.3a.1   \n",
       "81       A/California/81/2023  0.039266  2023.628767          2a.3a.1   \n",
       "82  A/SouthAfrica/R05384/2023  0.037561  2023.371233               2b   \n",
       "\n",
       "     subclade                 haplotype     method  \n",
       "0     G.1.1.2                   G.1.1.2  six_month  \n",
       "1     G.1.1.2              G.1.1.2:326R  six_month  \n",
       "2   G.1.3.1.1                 G.1.3.1.1  six_month  \n",
       "3   G.1.3.1.1            G.1.3.1.1:122D  six_month  \n",
       "4   G.1.3.1.1  G.1.3.1.1:122D,144N,276E  six_month  \n",
       "..        ...                       ...        ...  \n",
       "78  G.1.3.1.1              G.1.3.1.1:5E  six_month  \n",
       "79  G.1.3.1.1              G.1.3.1.1:5V  six_month  \n",
       "80  G.1.3.1.1             G.1.3.1.1:78D  six_month  \n",
       "81  G.1.3.1.1             G.1.3.1.1:81D  six_month  \n",
       "82        G.2        G.2:173L,212T,242M  six_month  \n",
       "\n",
       "[83 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subselect original dataframe on overlapping haplotypes\n",
    "overlapping_haplotypes_df = (strains_df[strains_df['haplotype']\n",
    "    .isin(overlapping_haplotypes)]\n",
    "    .query('method == \"six_month\"'))\n",
    "# Save list of overlapping haplotypes\n",
    "file = '6m_2y_overlapping_haplotypes.csv'\n",
    "overlapping_haplotypes_df.to_csv(os.path.join(resultsdir, file))\n",
    "print('Here are the overlapping haplotypes:')\n",
    "print(overlapping_haplotypes_df['haplotype'].tolist())\n",
    "\n",
    "# Subselect original dataframe on unique haplotypes\n",
    "unique_haplotypes_df = (strains_df[strains_df['haplotype']\n",
    "    .isin(unique_haplotypes)])\n",
    "\n",
    "\n",
    "# Add overlapping and unique dataframes back together\n",
    "auspice_strains_df = pd.concat([overlapping_haplotypes_df, unique_haplotypes_df]).reset_index(drop=True)\n",
    "print('\\nNow we can compare the number of strains from 6-month and 2-year, pre and post filtering.')\n",
    "print('We should have kept all 6 month strains, and only a subselection of 2-year, if there was overlap.')\n",
    "print('\\nHere is the input:')\n",
    "print('initial 6 month strains = ', len(strains_df.query('method == \"six_month\"')))\n",
    "print('initial 2 year strains = ', len(strains_df.query('method == \"two_year\"')))\n",
    "print('\\nAfter filtering:')\n",
    "print('the 6 month strains remaining = ', len(auspice_strains_df.query('method == \"six_month\"')))\n",
    "print('the 2 year strains remaining = ', len(auspice_strains_df.query('method == \"two_year\"')))\n",
    "print('therefore, the number of unique [Phylogenetic method] strains = ', \n",
    "      len(auspice_strains_df.query('method == \"six_month\"')) \n",
    "      + len(auspice_strains_df.query('method == \"two_year\"')))\n",
    "\n",
    "auspice_strains_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ec3fd1c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqName</th>\n",
       "      <th>short-clade</th>\n",
       "      <th>subclade</th>\n",
       "      <th>aaSubstitutions</th>\n",
       "      <th>HA1_substitutions</th>\n",
       "      <th>accession_ha</th>\n",
       "      <th>date</th>\n",
       "      <th>date_submitted</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>passage_category</th>\n",
       "      <th>originating_lab</th>\n",
       "      <th>submitting_lab</th>\n",
       "      <th>HA1_haplotype_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Portugal/210259/2022</td>\n",
       "      <td>2a.1</td>\n",
       "      <td>G.1.1</td>\n",
       "      <td>HA1:D104G,HA1:K276R</td>\n",
       "      <td>HA1:D104G,HA1:K276R</td>\n",
       "      <td>EPI2139262</td>\n",
       "      <td>2022-02-17</td>\n",
       "      <td>2022-08-31</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Portugal</td>\n",
       "      <td>Portugal</td>\n",
       "      <td>Portugal</td>\n",
       "      <td>cell</td>\n",
       "      <td>Instituto Nacional De Saude (insa)</td>\n",
       "      <td>Crick Worldwide Influenza Centre</td>\n",
       "      <td>4423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Ontario/12/2022</td>\n",
       "      <td>2a.1b</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K</td>\n",
       "      <td>HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K</td>\n",
       "      <td>EPI2337996</td>\n",
       "      <td>2022-11-09</td>\n",
       "      <td>2023-02-01</td>\n",
       "      <td>North America</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>unpassaged</td>\n",
       "      <td>Public Health Ontario</td>\n",
       "      <td>B.c. Centre For Disease Control</td>\n",
       "      <td>2708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A/Oklahoma/13632/2022</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H</td>\n",
       "      <td>EPI2290282</td>\n",
       "      <td>2022-11-17</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>North America</td>\n",
       "      <td>Usa</td>\n",
       "      <td>Oklahoma</td>\n",
       "      <td>Oklahoma</td>\n",
       "      <td>unpassaged</td>\n",
       "      <td>U.s. Air Force School Of Aerospace Medicine</td>\n",
       "      <td>U.s. Air Force School Of Aerospace Medicine</td>\n",
       "      <td>2124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A/Madrid/230625/2022</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2.2</td>\n",
       "      <td>HA1:R33Q,HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,...</td>\n",
       "      <td>HA1:R33Q,HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,...</td>\n",
       "      <td>EPI2478169</td>\n",
       "      <td>2022-12-14</td>\n",
       "      <td>2023-03-27</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Spain</td>\n",
       "      <td>Madrid</td>\n",
       "      <td>Madrid</td>\n",
       "      <td>unpassaged</td>\n",
       "      <td>Instituto De Salud Carlos Iii</td>\n",
       "      <td>Instituto De Salud Carlos Iii</td>\n",
       "      <td>1882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/Roraima/2022-020834-IEC/2022</td>\n",
       "      <td>2a.3</td>\n",
       "      <td>G.1.3</td>\n",
       "      <td>HA1:G53N,HA1:N96S,HA1:I192F,HA2:N49S,HA2:N160T</td>\n",
       "      <td>HA1:G53N,HA1:N96S,HA1:I192F</td>\n",
       "      <td>EPI2438576</td>\n",
       "      <td>2022-09-27</td>\n",
       "      <td>2023-03-03</td>\n",
       "      <td>South America</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Roraima</td>\n",
       "      <td>Roraima</td>\n",
       "      <td>unpassaged</td>\n",
       "      <td>Evandro Chagas Institute</td>\n",
       "      <td>Evandro Chagas Institute</td>\n",
       "      <td>1642</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          seqName short-clade subclade  \\\n",
       "0          A/Portugal/210259/2022        2a.1    G.1.1   \n",
       "1               A/Ontario/12/2022       2a.1b  G.1.1.2   \n",
       "2           A/Oklahoma/13632/2022          2b      G.2   \n",
       "3            A/Madrid/230625/2022          2b    G.2.2   \n",
       "4  A/Roraima/2022-020834-IEC/2022        2a.3    G.1.3   \n",
       "\n",
       "                                     aaSubstitutions  \\\n",
       "0                                HA1:D104G,HA1:K276R   \n",
       "1            HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K   \n",
       "2     HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H   \n",
       "3  HA1:R33Q,HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,...   \n",
       "4     HA1:G53N,HA1:N96S,HA1:I192F,HA2:N49S,HA2:N160T   \n",
       "\n",
       "                                   HA1_substitutions accession_ha        date  \\\n",
       "0                                HA1:D104G,HA1:K276R   EPI2139262  2022-02-17   \n",
       "1            HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K   EPI2337996  2022-11-09   \n",
       "2     HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H   EPI2290282  2022-11-17   \n",
       "3  HA1:R33Q,HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,...   EPI2478169  2022-12-14   \n",
       "4                        HA1:G53N,HA1:N96S,HA1:I192F   EPI2438576  2022-09-27   \n",
       "\n",
       "  date_submitted         region   country  division  location  \\\n",
       "0     2022-08-31         Europe  Portugal  Portugal  Portugal   \n",
       "1     2023-02-01  North America    Canada   Ontario   Ontario   \n",
       "2     2023-01-12  North America       Usa  Oklahoma  Oklahoma   \n",
       "3     2023-03-27         Europe     Spain    Madrid    Madrid   \n",
       "4     2023-03-03  South America    Brazil   Roraima   Roraima   \n",
       "\n",
       "  passage_category                               originating_lab  \\\n",
       "0             cell            Instituto Nacional De Saude (insa)   \n",
       "1       unpassaged                         Public Health Ontario   \n",
       "2       unpassaged  U.s. Air Force School Of Aerospace Medicine    \n",
       "3       unpassaged                 Instituto De Salud Carlos Iii   \n",
       "4       unpassaged                      Evandro Chagas Institute   \n",
       "\n",
       "                                submitting_lab  HA1_haplotype_count  \n",
       "0             Crick Worldwide Influenza Centre                 4423  \n",
       "1              B.c. Centre For Disease Control                 2708  \n",
       "2  U.s. Air Force School Of Aerospace Medicine                 2124  \n",
       "3                Instituto De Salud Carlos Iii                 1882  \n",
       "4                     Evandro Chagas Institute                 1642  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# [NextClade method]\n",
    "# My input is a single TSV, generated with help from John Huddleston (Bedford lab).\n",
    "# All outputs were generated from a metadata file generated on November 21st by John.\n",
    "\n",
    "metafile = 'representative_strains_per_HA1_haplotype_nextclade.tsv'\n",
    "\n",
    "nextclade_strains_df = pd.DataFrame()\n",
    "# Import emerging clades output\n",
    "df = pd.read_csv(os.path.join(datadir,'nextclade_haplotypes', metafile), sep='\\t')\n",
    "df['rankClass'] = key\n",
    "# Select minimum branch length strains\n",
    "nextclade_strains_df = pd.concat([nextclade_strains_df, df[['seqName', 'short-clade', 'subclade',\n",
    "                                                            'aaSubstitutions', 'HA1_substitutions', \n",
    "                                                            'accession_ha', 'date', 'date_submitted', \n",
    "                                                            'region', 'country', 'division', 'location', \n",
    "                                                            'passage_category', 'originating_lab', \n",
    "                                                            'submitting_lab', 'HA1_haplotype_count']]])\n",
    "\n",
    "\n",
    "nextclade_strains_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d855a0e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "here is an array of unique haplotype counts in 2022-2023...\n",
      "[4423 2708 2124 1882 1642 1626 1503  929  665  442  396  355  334  290\n",
      "  281  274  255  247  208  166  164  158  140  139  138  120  102   92\n",
      "   91   90   88   87   85   84   82   81   80   79   70   67   63   58\n",
      "   57   56   53   52   51   50   47   45   44   43   42   40   38   37\n",
      "   36   35   34   33   32   31   30   29   28   27   26   25   24   23\n",
      "   22   21   20   19   18   17   16   15   14   13   12   11   10    9\n",
      "    8    7    6    5    4    3    2    1]\n",
      "and here is a sample of the metadata file with counts...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqName</th>\n",
       "      <th>short-clade</th>\n",
       "      <th>subclade</th>\n",
       "      <th>aaSubstitutions</th>\n",
       "      <th>HA1_substitutions</th>\n",
       "      <th>accession_ha</th>\n",
       "      <th>date</th>\n",
       "      <th>date_submitted</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>passage_category</th>\n",
       "      <th>originating_lab</th>\n",
       "      <th>submitting_lab</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>accession_na</th>\n",
       "      <th>year</th>\n",
       "      <th>HA1_haplotype_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Catalonia/NSVH198262131/2022</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>H</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "      <td>EPI2260033</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>2022-12-23</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Spain</td>\n",
       "      <td>Catalonia</td>\n",
       "      <td>Catalonia</td>\n",
       "      <td>unpassaged</td>\n",
       "      <td>Hospital Universitari Germans Trias I Pujol</td>\n",
       "      <td>Hospital Universitari Vall D'hebron</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>EPI2260032</td>\n",
       "      <td>2022</td>\n",
       "      <td>1626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Portugal/210259/2022</td>\n",
       "      <td>2a.1</td>\n",
       "      <td>G.1.1</td>\n",
       "      <td>HA1:D104G,HA1:K276R</td>\n",
       "      <td>HA1:D104G,HA1:K276R</td>\n",
       "      <td>EPI2139262</td>\n",
       "      <td>2022-02-17</td>\n",
       "      <td>2022-08-31</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Portugal</td>\n",
       "      <td>Portugal</td>\n",
       "      <td>Portugal</td>\n",
       "      <td>cell</td>\n",
       "      <td>Instituto Nacional De Saude (insa)</td>\n",
       "      <td>Crick Worldwide Influenza Centre</td>\n",
       "      <td>48y</td>\n",
       "      <td>female</td>\n",
       "      <td>EPI2139347</td>\n",
       "      <td>2022</td>\n",
       "      <td>4423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A/NewJersey/13618/2022</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2.1</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,HA1:I140K...</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,HA1:I140K...</td>\n",
       "      <td>EPI2290191</td>\n",
       "      <td>2022-11-10</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>North America</td>\n",
       "      <td>Usa</td>\n",
       "      <td>New Jersey</td>\n",
       "      <td>New Jersey</td>\n",
       "      <td>unpassaged</td>\n",
       "      <td>U.s. Air Force School Of Aerospace Medicine</td>\n",
       "      <td>U.s. Air Force School Of Aerospace Medicine</td>\n",
       "      <td>34y</td>\n",
       "      <td>male</td>\n",
       "      <td>EPI2290190</td>\n",
       "      <td>2022</td>\n",
       "      <td>1503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A/Oklahoma/13632/2022</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H</td>\n",
       "      <td>EPI2290282</td>\n",
       "      <td>2022-11-17</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>North America</td>\n",
       "      <td>Usa</td>\n",
       "      <td>Oklahoma</td>\n",
       "      <td>Oklahoma</td>\n",
       "      <td>unpassaged</td>\n",
       "      <td>U.s. Air Force School Of Aerospace Medicine</td>\n",
       "      <td>U.s. Air Force School Of Aerospace Medicine</td>\n",
       "      <td>30y</td>\n",
       "      <td>male</td>\n",
       "      <td>EPI2290281</td>\n",
       "      <td>2022</td>\n",
       "      <td>2124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/Brazil/BA-LACEN-BA113-292060529/2022</td>\n",
       "      <td>2a.3</td>\n",
       "      <td>G.1.3</td>\n",
       "      <td>HA1:G53N,HA1:N96S,HA1:Y100C,HA1:I192F,HA2:N49S...</td>\n",
       "      <td>HA1:G53N,HA1:N96S,HA1:Y100C,HA1:I192F</td>\n",
       "      <td>EPI2185964</td>\n",
       "      <td>2022-01-06</td>\n",
       "      <td>2022-09-27</td>\n",
       "      <td>South America</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>Laboratório Central De Saúde Pública Professor...</td>\n",
       "      <td>Laborat?rio Central De Sa?de Publica Do Estado...</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>EPI2185963</td>\n",
       "      <td>2022</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  seqName short-clade subclade  \\\n",
       "0          A/Catalonia/NSVH198262131/2022     2a.3a.1        H   \n",
       "1                  A/Portugal/210259/2022        2a.1    G.1.1   \n",
       "2                  A/NewJersey/13618/2022          2b    G.2.1   \n",
       "3                   A/Oklahoma/13632/2022          2b      G.2   \n",
       "4  A/Brazil/BA-LACEN-BA113-292060529/2022        2a.3    G.1.3   \n",
       "\n",
       "                                     aaSubstitutions  \\\n",
       "0  HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...   \n",
       "1                                HA1:D104G,HA1:K276R   \n",
       "2  HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,HA1:I140K...   \n",
       "3     HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H   \n",
       "4  HA1:G53N,HA1:N96S,HA1:Y100C,HA1:I192F,HA2:N49S...   \n",
       "\n",
       "                                   HA1_substitutions accession_ha        date  \\\n",
       "0  HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...   EPI2260033  2022-12-15   \n",
       "1                                HA1:D104G,HA1:K276R   EPI2139262  2022-02-17   \n",
       "2  HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,HA1:I140K...   EPI2290191  2022-11-10   \n",
       "3     HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H   EPI2290282  2022-11-17   \n",
       "4              HA1:G53N,HA1:N96S,HA1:Y100C,HA1:I192F   EPI2185964  2022-01-06   \n",
       "\n",
       "  date_submitted         region   country    division    location  \\\n",
       "0     2022-12-23         Europe     Spain   Catalonia   Catalonia   \n",
       "1     2022-08-31         Europe  Portugal    Portugal    Portugal   \n",
       "2     2023-01-12  North America       Usa  New Jersey  New Jersey   \n",
       "3     2023-01-12  North America       Usa    Oklahoma    Oklahoma   \n",
       "4     2022-09-27  South America    Brazil      Brazil      Brazil   \n",
       "\n",
       "  passage_category                                    originating_lab  \\\n",
       "0       unpassaged       Hospital Universitari Germans Trias I Pujol    \n",
       "1             cell                 Instituto Nacional De Saude (insa)   \n",
       "2       unpassaged       U.s. Air Force School Of Aerospace Medicine    \n",
       "3       unpassaged       U.s. Air Force School Of Aerospace Medicine    \n",
       "4     undetermined  Laboratório Central De Saúde Pública Professor...   \n",
       "\n",
       "                                      submitting_lab  age  gender  \\\n",
       "0                Hospital Universitari Vall D'hebron    ?       ?   \n",
       "1                   Crick Worldwide Influenza Centre  48y  female   \n",
       "2        U.s. Air Force School Of Aerospace Medicine  34y    male   \n",
       "3        U.s. Air Force School Of Aerospace Medicine  30y    male   \n",
       "4  Laborat?rio Central De Sa?de Publica Do Estado...    ?       ?   \n",
       "\n",
       "  accession_na  year  HA1_haplotype_count  \n",
       "0   EPI2260032  2022                 1626  \n",
       "1   EPI2139347  2022                 4423  \n",
       "2   EPI2290190  2022                 1503  \n",
       "3   EPI2290281  2022                 2124  \n",
       "4   EPI2185963  2022                    1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge haplotype years and counts with metadata\n",
    "metadata_file = 'nextclade_metadata_h3n2_2023-11-21.tsv'\n",
    "metadata = pd.read_csv(os.path.join(datadir, metadata_file), sep='\\t')\n",
    "\n",
    "# Rename columns and add year\n",
    "metadata = (metadata.assign(year = lambda x: x['date'].str.split('-').str[0].astype(int))\n",
    "           .rename(columns={'name': 'seqName'})\n",
    "           )\n",
    "\n",
    "# Merge counts\n",
    "counted_metadata = metadata.merge(nextclade_strains_df[['HA1_substitutions', 'HA1_haplotype_count']])\n",
    "print('here is an array of unique haplotype counts in 2022-2023...')\n",
    "print(counted_metadata['HA1_haplotype_count'].sort_values(ascending=False).unique())\n",
    "print('and here is a sample of the metadata file with counts...')\n",
    "counted_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d8e718ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of 2022 records =  28790\n",
      "number of 2023 records =  5552\n",
      "here are the sampled clades : ['2a.3a.1' '2a.1' '2b' '2a.3' '1a.1' '2a.1a' '2a.1b' '2a.3a' '2a.2' '2a'\n",
      " '3C.2a1b.1a' '2c' '3C.2a1b.1b' '2a.3b' '2' '1' '3C.3a']\n",
      "and here is an array of unique haplotype counts only in 2023...\n",
      "[4423 2708 2124 1882 1642 1626 1503  665  442  396  355  334  281  274\n",
      "  255  247  208  164  158  139  138  102   91   90   88   82   81   80\n",
      "   79   70   63   58   53   51   50   47   44   43   40   36   33   31\n",
      "   30   29   28   27   26   25   24   23   22   21   20   19   18   17\n",
      "   16   15   14   13   12   11   10    9    8    7    6    5    4    3\n",
      "    2    1]\n"
     ]
    }
   ],
   "source": [
    "# Remove pre-2022 records\n",
    "print('number of 2022 records = ', len(counted_metadata.query('year == 2022')))\n",
    "print('number of 2023 records = ',len(counted_metadata.query('year == 2023')))\n",
    "\n",
    "recent_counted_metadata = (counted_metadata\n",
    "                           .query('year == 2023')\n",
    "                           .reset_index(drop=True)\n",
    "                           .rename(columns={'seqName': 'name'})\n",
    "                          )\n",
    "\n",
    "print('here are the sampled clades :', counted_metadata['short-clade'].unique())\n",
    "print('and here is an array of unique haplotype counts only in 2023...')\n",
    "print(recent_counted_metadata['HA1_haplotype_count'].sort_values(ascending=False).unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "679c1401",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "writing merged metadata to ../results/auspice_haplotypes_with_nextclade_counts.csv...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>div</th>\n",
       "      <th>num_date</th>\n",
       "      <th>clade_membership</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>method</th>\n",
       "      <th>subclade</th>\n",
       "      <th>date</th>\n",
       "      <th>date_submitted</th>\n",
       "      <th>HA1_substitutions</th>\n",
       "      <th>accession_ha</th>\n",
       "      <th>HA1_haplotype_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Finland/391/2023</td>\n",
       "      <td>0.033588</td>\n",
       "      <td>2023.513699</td>\n",
       "      <td>2a.1b</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>six_month</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>2023-07-07</td>\n",
       "      <td>2023-09-08</td>\n",
       "      <td>HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K</td>\n",
       "      <td>EPI2736643</td>\n",
       "      <td>2708.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Kanagawa/IC2239/2023</td>\n",
       "      <td>0.034155</td>\n",
       "      <td>2023.486301</td>\n",
       "      <td>2a.1b</td>\n",
       "      <td>G.1.1.2:326R</td>\n",
       "      <td>six_month</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>2023-06-27</td>\n",
       "      <td>2023-07-25</td>\n",
       "      <td>HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K,HA1:K326R</td>\n",
       "      <td>EPI2647253</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A/Singapore/NUH0526/2023</td>\n",
       "      <td>0.038133</td>\n",
       "      <td>2023.450685</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-06-14</td>\n",
       "      <td>2023-07-10</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "      <td>EPI2617594</td>\n",
       "      <td>1626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A/Sydney/332/2023</td>\n",
       "      <td>0.039833</td>\n",
       "      <td>2023.450685</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:122D</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-06-14</td>\n",
       "      <td>2023-07-24</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...</td>\n",
       "      <td>EPI2643073</td>\n",
       "      <td>274.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/Bangkok/P3755/2023</td>\n",
       "      <td>0.040966</td>\n",
       "      <td>2023.683562</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:122D,144N,276E</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-09-07</td>\n",
       "      <td>2023-09-28</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...</td>\n",
       "      <td>EPI2760070</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A/Ontario/RV00796/2023</td>\n",
       "      <td>0.040400</td>\n",
       "      <td>2023.563014</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:122D,276E</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-07-25</td>\n",
       "      <td>2023-09-07</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...</td>\n",
       "      <td>EPI2735889</td>\n",
       "      <td>70.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A/Krabi/THIS050/2023</td>\n",
       "      <td>0.042670</td>\n",
       "      <td>2023.645205</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:124N,145N</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-08-24</td>\n",
       "      <td>2023-09-28</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:S124N,HA1:I140K...</td>\n",
       "      <td>EPI2759946</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A/Saskatchewan/SKFLU317847/2023</td>\n",
       "      <td>0.039833</td>\n",
       "      <td>2023.617808</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:145N</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-08-14</td>\n",
       "      <td>2023-09-11</td>\n",
       "      <td>HA1:T28A,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,...</td>\n",
       "      <td>EPI2743304</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A/StPetersburg/MH144113/2023</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>2023.626027</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:173R,276E</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-08-17</td>\n",
       "      <td>2023-10-24</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:Q173R...</td>\n",
       "      <td>EPI2777879</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A/Maldives/852/2023</td>\n",
       "      <td>0.039266</td>\n",
       "      <td>2023.420548</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:21T</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-06-03</td>\n",
       "      <td>2023-08-17</td>\n",
       "      <td>HA1:P21T,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,...</td>\n",
       "      <td>EPI2689677</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A/Georgia/40/2023</td>\n",
       "      <td>0.039266</td>\n",
       "      <td>2023.642466</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:242M</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-08-23</td>\n",
       "      <td>2023-09-25</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "      <td>EPI2756864</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A/Bhutan/845/2023</td>\n",
       "      <td>0.040402</td>\n",
       "      <td>2023.546575</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:25V</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-07-19</td>\n",
       "      <td>2023-10-31</td>\n",
       "      <td>HA1:I25V,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,...</td>\n",
       "      <td>EPI2781036</td>\n",
       "      <td>139.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A/Catalonia/2041146NS/2023</td>\n",
       "      <td>0.035307</td>\n",
       "      <td>2023.393151</td>\n",
       "      <td>2a.3b</td>\n",
       "      <td>G.1.3.2</td>\n",
       "      <td>six_month</td>\n",
       "      <td>G.1.3.2</td>\n",
       "      <td>2023-05-24</td>\n",
       "      <td>2023-10-24</td>\n",
       "      <td>HA1:G53N,HA1:N96S,HA1:I140M,HA1:I192F</td>\n",
       "      <td>EPI2778022</td>\n",
       "      <td>355.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A/Jeju/1047/2023</td>\n",
       "      <td>0.033589</td>\n",
       "      <td>2023.373973</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2</td>\n",
       "      <td>six_month</td>\n",
       "      <td>G.2</td>\n",
       "      <td>2023-05-17</td>\n",
       "      <td>2023-09-07</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H</td>\n",
       "      <td>EPI2735103</td>\n",
       "      <td>2124.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A/Kobe/23117/2023</td>\n",
       "      <td>0.034156</td>\n",
       "      <td>2023.546575</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2.1</td>\n",
       "      <td>six_month</td>\n",
       "      <td>G.2.1</td>\n",
       "      <td>2023-07-19</td>\n",
       "      <td>2023-09-26</td>\n",
       "      <td>HA1:G5E,HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,H...</td>\n",
       "      <td>EPI2758716</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A/Tecpan/17FLU/2023</td>\n",
       "      <td>0.034156</td>\n",
       "      <td>2023.373973</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2:242M</td>\n",
       "      <td>six_month</td>\n",
       "      <td>G.2</td>\n",
       "      <td>2023-05-17</td>\n",
       "      <td>2023-11-14</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H...</td>\n",
       "      <td>EPI2791811</td>\n",
       "      <td>665.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A/Wisconsin/27/2023</td>\n",
       "      <td>0.046176</td>\n",
       "      <td>2023.176712</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:309I</td>\n",
       "      <td>two_year</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-03-06</td>\n",
       "      <td>2023-04-04</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "      <td>EPI2498763</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A/Pakistan/GIHSN/ICT/1339/2023</td>\n",
       "      <td>0.043907</td>\n",
       "      <td>2023.182192</td>\n",
       "      <td>2a.3b</td>\n",
       "      <td>G.1.3.2:83K</td>\n",
       "      <td>two_year</td>\n",
       "      <td>G.1.3.2</td>\n",
       "      <td>2023-03-08</td>\n",
       "      <td>2023-03-23</td>\n",
       "      <td>HA1:G53N,HA1:E83K,HA1:N96S,HA1:I140M,HA1:I192F</td>\n",
       "      <td>EPI2505314</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A/Bhutan/6/2023</td>\n",
       "      <td>0.048463</td>\n",
       "      <td>2023.182192</td>\n",
       "      <td>2a.3</td>\n",
       "      <td>G.1.3:48T,173R</td>\n",
       "      <td>two_year</td>\n",
       "      <td>G.1.3</td>\n",
       "      <td>2023-03-08</td>\n",
       "      <td>2023-05-23</td>\n",
       "      <td>HA1:I48T,HA1:G53N,HA1:N96S,HA1:Q173R,HA1:I192F</td>\n",
       "      <td>EPI2571554</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A/Luga/RII-11393S/2023</td>\n",
       "      <td>0.042197</td>\n",
       "      <td>2023.100000</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2.1:269K</td>\n",
       "      <td>two_year</td>\n",
       "      <td>G.2.1</td>\n",
       "      <td>2023-02-06</td>\n",
       "      <td>2023-05-17</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,HA1:I140K...</td>\n",
       "      <td>EPI2558520</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A/SouthDakota/22/2023</td>\n",
       "      <td>0.041062</td>\n",
       "      <td>2023.149315</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2:101E</td>\n",
       "      <td>two_year</td>\n",
       "      <td>G.2</td>\n",
       "      <td>2023-02-24</td>\n",
       "      <td>2023-04-04</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:D101E,HA1:I140K...</td>\n",
       "      <td>EPI2498814</td>\n",
       "      <td>247.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A/SuratThani/P152/2023</td>\n",
       "      <td>0.041631</td>\n",
       "      <td>2023.031507</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2:7G</td>\n",
       "      <td>two_year</td>\n",
       "      <td>G.2</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>2023-01-25</td>\n",
       "      <td>HA1:D7G,HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,H...</td>\n",
       "      <td>EPI2312127</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A/Busan/1301/2023</td>\n",
       "      <td>0.034723</td>\n",
       "      <td>2023.563014</td>\n",
       "      <td>2a.1</td>\n",
       "      <td>B</td>\n",
       "      <td>six_month</td>\n",
       "      <td>G.1.1</td>\n",
       "      <td>2023-07-25</td>\n",
       "      <td>2023-11-15</td>\n",
       "      <td>HA1:K276R</td>\n",
       "      <td>EPI2793114</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A/Sydney/715/2023</td>\n",
       "      <td>0.041533</td>\n",
       "      <td>2023.582192</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:104N</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-08-01</td>\n",
       "      <td>2023-09-04</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:D104N,HA1:I140K...</td>\n",
       "      <td>EPI2724875</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A/Sydney/749/2023</td>\n",
       "      <td>0.039267</td>\n",
       "      <td>2023.598630</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:106V</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-08-07</td>\n",
       "      <td>2023-11-06</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:A106V,HA1:I140K...</td>\n",
       "      <td>EPI2785896</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A/Netherlands/1693/2023</td>\n",
       "      <td>0.040966</td>\n",
       "      <td>2023.700000</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:122D,276E,327R</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-09-13</td>\n",
       "      <td>2023-10-11</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...</td>\n",
       "      <td>EPI2771080</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A/SouthAfrica/R07188/2023</td>\n",
       "      <td>0.039833</td>\n",
       "      <td>2023.469863</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:124N</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-06-21</td>\n",
       "      <td>2023-09-04</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:S124N,HA1:I140K...</td>\n",
       "      <td>EPI2725146</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A/SamutSongkhram/THIS061/2023</td>\n",
       "      <td>0.041534</td>\n",
       "      <td>2023.664384</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:197H</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-08-31</td>\n",
       "      <td>2023-09-28</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "      <td>EPI2759881</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A/Chipata/15-NIC-001/2023</td>\n",
       "      <td>0.039832</td>\n",
       "      <td>2023.565753</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:198P,207R</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-07-26</td>\n",
       "      <td>2023-10-01</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "      <td>EPI2762684</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A/Solwezi/13-NIC-001/2023</td>\n",
       "      <td>0.041534</td>\n",
       "      <td>2023.541096</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:21T,122D</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-07-17</td>\n",
       "      <td>2023-10-12</td>\n",
       "      <td>HA1:P21T,HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,...</td>\n",
       "      <td>EPI2773303</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>A/Victoria/1033/2023</td>\n",
       "      <td>0.038699</td>\n",
       "      <td>2023.401370</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:278E</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-05-27</td>\n",
       "      <td>2023-08-15</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "      <td>EPI2686112</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A/SouthAfrica/R05510/2023</td>\n",
       "      <td>0.040399</td>\n",
       "      <td>2023.387671</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:299K</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-05-22</td>\n",
       "      <td>2023-08-16</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "      <td>EPI2687991</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>A/Netherlands/1685/2023</td>\n",
       "      <td>0.040403</td>\n",
       "      <td>2023.686301</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:5E</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-09-08</td>\n",
       "      <td>2023-09-22</td>\n",
       "      <td>HA1:G5E,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,H...</td>\n",
       "      <td>EPI2755805</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>A/SouthAfrica/R06477/2023</td>\n",
       "      <td>0.039833</td>\n",
       "      <td>2023.417808</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:5V</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-06-02</td>\n",
       "      <td>2023-08-16</td>\n",
       "      <td>HA1:G5V,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,H...</td>\n",
       "      <td>EPI2688989</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>A/Oman/3011/2023</td>\n",
       "      <td>0.039266</td>\n",
       "      <td>2023.543836</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:78D</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-07-18</td>\n",
       "      <td>2023-09-25</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:G78D,HA1:N96S,HA1:I140K,...</td>\n",
       "      <td>EPI2756778</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>A/California/81/2023</td>\n",
       "      <td>0.039266</td>\n",
       "      <td>2023.628767</td>\n",
       "      <td>2a.3a.1</td>\n",
       "      <td>G.1.3.1.1:81D</td>\n",
       "      <td>six_month</td>\n",
       "      <td>H</td>\n",
       "      <td>2023-08-18</td>\n",
       "      <td>2023-10-11</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N81D,HA1:N96S,HA1:I140K,...</td>\n",
       "      <td>EPI2772585</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>A/SouthAfrica/R05384/2023</td>\n",
       "      <td>0.037561</td>\n",
       "      <td>2023.371233</td>\n",
       "      <td>2b</td>\n",
       "      <td>G.2:173L,212T,242M</td>\n",
       "      <td>six_month</td>\n",
       "      <td>G.2</td>\n",
       "      <td>2023-05-16</td>\n",
       "      <td>2023-07-26</td>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H...</td>\n",
       "      <td>EPI2653147</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               name       div     num_date clade_membership  \\\n",
       "0                A/Finland/391/2023  0.033588  2023.513699            2a.1b   \n",
       "1            A/Kanagawa/IC2239/2023  0.034155  2023.486301            2a.1b   \n",
       "2          A/Singapore/NUH0526/2023  0.038133  2023.450685          2a.3a.1   \n",
       "3                 A/Sydney/332/2023  0.039833  2023.450685          2a.3a.1   \n",
       "4              A/Bangkok/P3755/2023  0.040966  2023.683562          2a.3a.1   \n",
       "5            A/Ontario/RV00796/2023  0.040400  2023.563014          2a.3a.1   \n",
       "6              A/Krabi/THIS050/2023  0.042670  2023.645205          2a.3a.1   \n",
       "7   A/Saskatchewan/SKFLU317847/2023  0.039833  2023.617808          2a.3a.1   \n",
       "8      A/StPetersburg/MH144113/2023  0.042105  2023.626027          2a.3a.1   \n",
       "9               A/Maldives/852/2023  0.039266  2023.420548          2a.3a.1   \n",
       "10                A/Georgia/40/2023  0.039266  2023.642466          2a.3a.1   \n",
       "11                A/Bhutan/845/2023  0.040402  2023.546575          2a.3a.1   \n",
       "12       A/Catalonia/2041146NS/2023  0.035307  2023.393151            2a.3b   \n",
       "13                 A/Jeju/1047/2023  0.033589  2023.373973               2b   \n",
       "14                A/Kobe/23117/2023  0.034156  2023.546575               2b   \n",
       "15              A/Tecpan/17FLU/2023  0.034156  2023.373973               2b   \n",
       "16              A/Wisconsin/27/2023  0.046176  2023.176712          2a.3a.1   \n",
       "17   A/Pakistan/GIHSN/ICT/1339/2023  0.043907  2023.182192            2a.3b   \n",
       "18                  A/Bhutan/6/2023  0.048463  2023.182192             2a.3   \n",
       "19           A/Luga/RII-11393S/2023  0.042197  2023.100000               2b   \n",
       "20            A/SouthDakota/22/2023  0.041062  2023.149315               2b   \n",
       "21           A/SuratThani/P152/2023  0.041631  2023.031507               2b   \n",
       "22                A/Busan/1301/2023  0.034723  2023.563014             2a.1   \n",
       "23                A/Sydney/715/2023  0.041533  2023.582192          2a.3a.1   \n",
       "24                A/Sydney/749/2023  0.039267  2023.598630          2a.3a.1   \n",
       "25          A/Netherlands/1693/2023  0.040966  2023.700000          2a.3a.1   \n",
       "26        A/SouthAfrica/R07188/2023  0.039833  2023.469863          2a.3a.1   \n",
       "27    A/SamutSongkhram/THIS061/2023  0.041534  2023.664384          2a.3a.1   \n",
       "28        A/Chipata/15-NIC-001/2023  0.039832  2023.565753          2a.3a.1   \n",
       "29        A/Solwezi/13-NIC-001/2023  0.041534  2023.541096          2a.3a.1   \n",
       "30             A/Victoria/1033/2023  0.038699  2023.401370          2a.3a.1   \n",
       "31        A/SouthAfrica/R05510/2023  0.040399  2023.387671          2a.3a.1   \n",
       "32          A/Netherlands/1685/2023  0.040403  2023.686301          2a.3a.1   \n",
       "33        A/SouthAfrica/R06477/2023  0.039833  2023.417808          2a.3a.1   \n",
       "34                 A/Oman/3011/2023  0.039266  2023.543836          2a.3a.1   \n",
       "35             A/California/81/2023  0.039266  2023.628767          2a.3a.1   \n",
       "36        A/SouthAfrica/R05384/2023  0.037561  2023.371233               2b   \n",
       "\n",
       "                   haplotype     method subclade        date date_submitted  \\\n",
       "0                    G.1.1.2  six_month  G.1.1.2  2023-07-07     2023-09-08   \n",
       "1               G.1.1.2:326R  six_month  G.1.1.2  2023-06-27     2023-07-25   \n",
       "2                  G.1.3.1.1  six_month        H  2023-06-14     2023-07-10   \n",
       "3             G.1.3.1.1:122D  six_month        H  2023-06-14     2023-07-24   \n",
       "4   G.1.3.1.1:122D,144N,276E  six_month        H  2023-09-07     2023-09-28   \n",
       "5        G.1.3.1.1:122D,276E  six_month        H  2023-07-25     2023-09-07   \n",
       "6        G.1.3.1.1:124N,145N  six_month        H  2023-08-24     2023-09-28   \n",
       "7             G.1.3.1.1:145N  six_month        H  2023-08-14     2023-09-11   \n",
       "8        G.1.3.1.1:173R,276E  six_month        H  2023-08-17     2023-10-24   \n",
       "9              G.1.3.1.1:21T  six_month        H  2023-06-03     2023-08-17   \n",
       "10            G.1.3.1.1:242M  six_month        H  2023-08-23     2023-09-25   \n",
       "11             G.1.3.1.1:25V  six_month        H  2023-07-19     2023-10-31   \n",
       "12                   G.1.3.2  six_month  G.1.3.2  2023-05-24     2023-10-24   \n",
       "13                       G.2  six_month      G.2  2023-05-17     2023-09-07   \n",
       "14                     G.2.1  six_month    G.2.1  2023-07-19     2023-09-26   \n",
       "15                  G.2:242M  six_month      G.2  2023-05-17     2023-11-14   \n",
       "16            G.1.3.1.1:309I   two_year        H  2023-03-06     2023-04-04   \n",
       "17               G.1.3.2:83K   two_year  G.1.3.2  2023-03-08     2023-03-23   \n",
       "18            G.1.3:48T,173R   two_year    G.1.3  2023-03-08     2023-05-23   \n",
       "19                G.2.1:269K   two_year    G.2.1  2023-02-06     2023-05-17   \n",
       "20                  G.2:101E   two_year      G.2  2023-02-24     2023-04-04   \n",
       "21                    G.2:7G   two_year      G.2  2023-01-12     2023-01-25   \n",
       "22                         B  six_month    G.1.1  2023-07-25     2023-11-15   \n",
       "23            G.1.3.1.1:104N  six_month        H  2023-08-01     2023-09-04   \n",
       "24            G.1.3.1.1:106V  six_month        H  2023-08-07     2023-11-06   \n",
       "25  G.1.3.1.1:122D,276E,327R  six_month        H  2023-09-13     2023-10-11   \n",
       "26            G.1.3.1.1:124N  six_month        H  2023-06-21     2023-09-04   \n",
       "27            G.1.3.1.1:197H  six_month        H  2023-08-31     2023-09-28   \n",
       "28       G.1.3.1.1:198P,207R  six_month        H  2023-07-26     2023-10-01   \n",
       "29        G.1.3.1.1:21T,122D  six_month        H  2023-07-17     2023-10-12   \n",
       "30            G.1.3.1.1:278E  six_month        H  2023-05-27     2023-08-15   \n",
       "31            G.1.3.1.1:299K  six_month        H  2023-05-22     2023-08-16   \n",
       "32              G.1.3.1.1:5E  six_month        H  2023-09-08     2023-09-22   \n",
       "33              G.1.3.1.1:5V  six_month        H  2023-06-02     2023-08-16   \n",
       "34             G.1.3.1.1:78D  six_month        H  2023-07-18     2023-09-25   \n",
       "35             G.1.3.1.1:81D  six_month        H  2023-08-18     2023-10-11   \n",
       "36        G.2:173L,212T,242M  six_month      G.2  2023-05-16     2023-07-26   \n",
       "\n",
       "                                    HA1_substitutions accession_ha  \\\n",
       "0             HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K   EPI2736643   \n",
       "1   HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K,HA1:K326R   EPI2647253   \n",
       "2   HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...   EPI2617594   \n",
       "3   HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...   EPI2643073   \n",
       "4   HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...   EPI2760070   \n",
       "5   HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...   EPI2735889   \n",
       "6   HA1:E50K,HA1:G53N,HA1:N96S,HA1:S124N,HA1:I140K...   EPI2759946   \n",
       "7   HA1:T28A,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,...   EPI2743304   \n",
       "8   HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:Q173R...   EPI2777879   \n",
       "9   HA1:P21T,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,...   EPI2689677   \n",
       "10  HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...   EPI2756864   \n",
       "11  HA1:I25V,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,...   EPI2781036   \n",
       "12              HA1:G53N,HA1:N96S,HA1:I140M,HA1:I192F   EPI2778022   \n",
       "13     HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H   EPI2735103   \n",
       "14  HA1:G5E,HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,H...   EPI2758716   \n",
       "15  HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H...   EPI2791811   \n",
       "16  HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...   EPI2498763   \n",
       "17     HA1:G53N,HA1:E83K,HA1:N96S,HA1:I140M,HA1:I192F   EPI2505314   \n",
       "18     HA1:I48T,HA1:G53N,HA1:N96S,HA1:Q173R,HA1:I192F   EPI2571554   \n",
       "19  HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,HA1:I140K...   EPI2558520   \n",
       "20  HA1:E50K,HA1:G53D,HA1:F79V,HA1:D101E,HA1:I140K...   EPI2498814   \n",
       "21  HA1:D7G,HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,H...   EPI2312127   \n",
       "22                                          HA1:K276R   EPI2793114   \n",
       "23  HA1:E50K,HA1:G53N,HA1:N96S,HA1:D104N,HA1:I140K...   EPI2724875   \n",
       "24  HA1:E50K,HA1:G53N,HA1:N96S,HA1:A106V,HA1:I140K...   EPI2785896   \n",
       "25  HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...   EPI2771080   \n",
       "26  HA1:E50K,HA1:G53N,HA1:N96S,HA1:S124N,HA1:I140K...   EPI2725146   \n",
       "27  HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...   EPI2759881   \n",
       "28  HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...   EPI2762684   \n",
       "29  HA1:P21T,HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,...   EPI2773303   \n",
       "30  HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...   EPI2686112   \n",
       "31  HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...   EPI2687991   \n",
       "32  HA1:G5E,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,H...   EPI2755805   \n",
       "33  HA1:G5V,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,H...   EPI2688989   \n",
       "34  HA1:E50K,HA1:G53N,HA1:G78D,HA1:N96S,HA1:I140K,...   EPI2756778   \n",
       "35  HA1:E50K,HA1:G53N,HA1:N81D,HA1:N96S,HA1:I140K,...   EPI2772585   \n",
       "36  HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H...   EPI2653147   \n",
       "\n",
       "    HA1_haplotype_count  \n",
       "0                2708.0  \n",
       "1                  11.0  \n",
       "2                1626.0  \n",
       "3                 274.0  \n",
       "4                  10.0  \n",
       "5                  70.0  \n",
       "6                  20.0  \n",
       "7                   1.0  \n",
       "8                  28.0  \n",
       "9                  18.0  \n",
       "10                 10.0  \n",
       "11                139.0  \n",
       "12                355.0  \n",
       "13               2124.0  \n",
       "14                 23.0  \n",
       "15                665.0  \n",
       "16                 16.0  \n",
       "17                 79.0  \n",
       "18                  5.0  \n",
       "19                 28.0  \n",
       "20                247.0  \n",
       "21                 10.0  \n",
       "22                 36.0  \n",
       "23                  9.0  \n",
       "24                  7.0  \n",
       "25                 19.0  \n",
       "26                 14.0  \n",
       "27                  1.0  \n",
       "28                  6.0  \n",
       "29                 15.0  \n",
       "30                 13.0  \n",
       "31                 15.0  \n",
       "32                  5.0  \n",
       "33                  6.0  \n",
       "34                 28.0  \n",
       "35                  7.0  \n",
       "36                  7.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now merge with auspice \n",
    "auspice_nextclade_merged_df = (auspice_strains_df\n",
    "                               .drop(columns = ['subclade'])\n",
    "    .merge(recent_counted_metadata[['name', 'subclade', 'date', 'date_submitted', 'HA1_substitutions', 'accession_ha', 'HA1_haplotype_count']], how='left', suffixes = ('_auspice2y', '_nextclade'), on = ['name'])\n",
    "    .dropna()\n",
    "    .reset_index(drop=True)\n",
    "    )\n",
    "# Metadata contains a weird typo where 'EPI' in HA accession is sometimes repeated\n",
    "auspice_nextclade_merged_df['accession_ha'] = auspice_nextclade_merged_df['accession_ha'].str.replace('EPIEPI','EPI')\n",
    "\n",
    "# Write out results\n",
    "outfile = os.path.join(resultsdir, 'auspice_haplotypes_with_nextclade_counts.csv')\n",
    "print(f'writing merged metadata to {outfile}...')\n",
    "auspice_nextclade_merged_df.to_csv(outfile, index=False)\n",
    "auspice_nextclade_merged_df\n",
    "# ['name', 'HA1_substitutions']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1501b67c",
   "metadata": {},
   "source": [
    "## Adding new HA1 subtitutions\n",
    "The above method only found 37 haplotypes. We have more room in this library and want to include new mutations.\n",
    "\n",
    "To do that, I'll identify all HA1 substitutions sampled within the last 6 months that have a 2022-2023 record count of greater than 5. \n",
    "I'll compare these haplotypes to my auspice-nextclade method, detailed above.\n",
    "I suspect I'll have overlap, but also identify new HA1 substitutions. \n",
    "\n",
    "To pick representative strains for any new HA1 substitutions, I'll choose the most recent sampled strain. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "64d34a4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "some of the month dates are input as \"X\", so we will have to handle those differently\n",
      "here are the 15 unsure 2023 HA1 substitutions:\n",
      "['HA1:I25V,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:R150K,HA1:I192F,HA1:I223V'\n",
      " 'HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F,HA1:I223V'\n",
      " 'HA1:R33Q,HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H,HA1:S262N'\n",
      " 'HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,HA1:I140K,HA1:S156H,HA1:S262N'\n",
      " 'HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F,HA1:I223V,HA1:I242M'\n",
      " 'HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H'\n",
      " 'HA1:I25V,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F,HA1:K207E,HA1:I223V'\n",
      " 'HA1:I25V,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F,HA1:I223V'\n",
      " 'HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K,HA1:I192F,HA1:I223V,HA1:K276E'\n",
      " 'HA1:E50K,HA1:G53N,HA1:N96R,HA1:I140K,HA1:I192F,HA1:I223V'\n",
      " 'HA1:I25V,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F,HA1:I223V,HA1:D271E'\n",
      " 'HA1:E50K,HA1:G53N,HA1:N96S,HA1:A106V,HA1:I140K,HA1:I192F,HA1:I223V'\n",
      " 'HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K,HA1:I192F,HA1:I223V'\n",
      " 'HA1:D7G,HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H'\n",
      " 'HA1:T30A,HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F,HA1:I223V']\n",
      "\n",
      "here are submitted dates for uncertain 2023 sequences\n",
      "['2023-09-18', '2023-09-11', '2023-03-15', '2023-08-08', '2023-06-15']\n",
      "\n",
      "the number of uncertain 2023 HA1 haplotypes SUBMITTED in past 6 months...\n",
      "15\n",
      "\n",
      "the number of uncertain 2023 HA1 haplotypes SUBMITTED in past 6 months with >= 5 counts...\n",
      "9\n",
      "\n",
      "the number of HA1 haplotypes in past 6 months...\n",
      "264\n",
      "\n",
      "the number of HA1 haplotypes in past 6 months with >= 5 counts...\n",
      "54\n"
     ]
    }
   ],
   "source": [
    "# Rename columns and add year\n",
    "counted_metadata = counted_metadata.assign(year_month_day = lambda x: x['date'].str.split('-').str[0:3])\n",
    "counted_metadata['yearmonthday'] = [''.join(map(str, l)) for l in counted_metadata['year_month_day']]\n",
    "counted_metadata = counted_metadata.drop(columns = ['year_month_day'])\n",
    "\n",
    "# Count uncertain dates\n",
    "print('some of the month dates are input as \"X\", so we will have to handle those differently')\n",
    "count = len(counted_metadata.query('yearmonthday == \"2023XXXX\"').HA1_substitutions.unique().tolist())\n",
    "print(f'here are the {count} unsure 2023 HA1 substitutions:')\n",
    "print(counted_metadata.query('yearmonthday == \"2023XXXX\"').HA1_substitutions.unique())\n",
    "\n",
    "\n",
    "# ID uncertain 2023 metadata and handle separately \n",
    "uncertain_2023_metadata = counted_metadata.query('yearmonthday == \"2023XXXX\"')\n",
    "print('\\nhere are submitted dates for uncertain 2023 sequences')\n",
    "print(uncertain_2023_metadata.date_submitted.unique().tolist())\n",
    "uncertain_2023_metadata = uncertain_2023_metadata.assign(\n",
    "    month = lambda x: x['date_submitted'].str.split('-').str[1].astype(int)\n",
    ")\n",
    "\n",
    "print('\\nthe number of uncertain 2023 HA1 haplotypes SUBMITTED in past 6 months...')\n",
    "uncertain_past_six = (uncertain_2023_metadata[uncertain_2023_metadata.month >= 6])\n",
    "print(len(uncertain_2023_metadata['HA1_substitutions'].unique().tolist()))\n",
    "\n",
    "print('\\nthe number of uncertain 2023 HA1 haplotypes SUBMITTED in past 6 months with >= 5 counts...')\n",
    "uncertain_greater_that_five = uncertain_past_six[uncertain_past_six.HA1_haplotype_count >= 5]\n",
    "print(len(uncertain_greater_that_five['HA1_substitutions'].unique().tolist()))\n",
    "\n",
    "\n",
    "# Subtract uncertain 2022 and 2023 dates from counted_metadata\n",
    "num_counted_metadata = counted_metadata[pd.to_numeric(counted_metadata['yearmonthday'], errors='coerce').notnull()]\n",
    "num_counted_metadata = num_counted_metadata.assign(\n",
    "    yearmonthday = lambda x: x['yearmonthday'].astype(int)\n",
    ")\n",
    "\n",
    "print('\\nthe number of HA1 haplotypes in past 6 months...')\n",
    "past_six = num_counted_metadata[num_counted_metadata.yearmonthday >= 20230600]\n",
    "print(len(past_six['HA1_substitutions'].unique().tolist()))\n",
    "\n",
    "print('\\nthe number of HA1 haplotypes in past 6 months with >= 5 counts...')\n",
    "greater_that_five = past_six[past_six.HA1_haplotype_count >= 5]\n",
    "print(len(greater_that_five['HA1_substitutions'].unique().tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "99bb8fe5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "here is a preview of the total unique haplotypes from the combined \"Auspice\" and \"New HA1 sub\" methods\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HA1_substitutions</th>\n",
       "      <th>HA1_haplotype_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...</td>\n",
       "      <td>70.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...</td>\n",
       "      <td>274.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:G78D,HA1:N96S,HA1:I140K,...</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:S145N...</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>HA1:I48T,HA1:G53N,HA1:N96S,HA1:Q173R,HA1:I192F</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,HA1:I140K...</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:D101E,HA1:I140K...</td>\n",
       "      <td>247.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H...</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    HA1_substitutions  HA1_haplotype_count\n",
       "0   HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...                 70.0\n",
       "1   HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...                274.0\n",
       "2   HA1:E50K,HA1:G53N,HA1:G78D,HA1:N96S,HA1:I140K,...                 28.0\n",
       "3   HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:S145N...                 23.0\n",
       "4   HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...                 10.0\n",
       "..                                                ...                  ...\n",
       "57     HA1:I48T,HA1:G53N,HA1:N96S,HA1:Q173R,HA1:I192F                  5.0\n",
       "58  HA1:E50K,HA1:G53D,HA1:F79V,HA1:T135A,HA1:I140K...                 28.0\n",
       "59  HA1:E50K,HA1:G53D,HA1:F79V,HA1:D101E,HA1:I140K...                247.0\n",
       "60  HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...                  1.0\n",
       "61  HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H...                  7.0\n",
       "\n",
       "[62 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('here is a preview of the total unique haplotypes from the combined \"Auspice\" and \"New HA1 sub\" methods')\n",
    "(pd.concat([greater_that_five.rename(columns={'seqName': 'name'}),\n",
    "            uncertain_greater_that_five.rename(columns={'seqName': 'name'}),\n",
    "            auspice_nextclade_merged_df])\n",
    " [['HA1_substitutions', 'HA1_haplotype_count']]\n",
    " .drop_duplicates()\n",
    " .reset_index(drop=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2c706907",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this method identifies 25 new HA1 substitutions...\n",
      "here are some the representative strains for each substitution, chosen by latest collection date...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['A/Sydney/710/2023',\n",
       " 'A/Netherlands/1760/2023',\n",
       " 'A/SouthAfrica/KO56863/2023',\n",
       " 'A/SouthAfrica/R06240/2023',\n",
       " 'A/Finland/399/2023',\n",
       " 'A/SouthAfrica/PET28931/2023',\n",
       " 'A/SouthAfrica/R06359/2023',\n",
       " 'A/SouthAfrica/R06506/2023',\n",
       " 'A/Sendai/45/2023',\n",
       " 'A/Townsville/68/2023',\n",
       " 'A/Bangkok/P3599/2023',\n",
       " 'A/SouthAfrica/K056872/2023',\n",
       " 'A/Okinawa/234/2023',\n",
       " 'A/Guangdong-Futian/1980/2023',\n",
       " 'A/SouthSudan/642/2023',\n",
       " 'A/Ehime/50/2023',\n",
       " 'A/Chipata/15-NIC-007/2023',\n",
       " 'A/SchleswigHolstein/4/2023',\n",
       " 'A/Catalonia/NSVH102124476/2023',\n",
       " 'A/AbuDhabi/6753/2023',\n",
       " 'A/Kanagawa/AC2316/2023',\n",
       " 'A/Brisbane/429/2023',\n",
       " 'A/SouthAfrica/R07073/2023',\n",
       " 'A/SouthSudan/631/2023',\n",
       " 'A/Southafrica/R07876/2020']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List of auspice-method HA1 mutations already represented in library\n",
    "auspice_HA1_muts = auspice_nextclade_merged_df['HA1_substitutions'].tolist()\n",
    "\n",
    "# Concat uncertain-2023 and 2023 dataframes\n",
    "all_HA1_substitutions = (pd.concat([greater_that_five.rename(columns={'seqName': 'name'}),\n",
    "                                    uncertain_greater_that_five.rename(columns={'seqName': 'name'})])\n",
    "                         .drop_duplicates()\n",
    "                         .drop(columns=['month'])\n",
    "                         .reset_index(drop=True)\n",
    "                        )\n",
    "\n",
    "new_subs_df = all_HA1_substitutions[~all_HA1_substitutions.HA1_substitutions.isin(auspice_HA1_muts)]\n",
    "\n",
    "# ID new unqiue HA1 substitutions\n",
    "new_HA_list = new_subs_df.HA1_substitutions.unique().tolist()\n",
    "print(f'this method identifies {len(new_HA_list)} new HA1 substitutions...')\n",
    "\n",
    "# Save representative strains with most recent collection date \n",
    "new_HA_strains_df = pd.DataFrame()\n",
    "\n",
    "for ha in new_HA_list:\n",
    "    df = new_subs_df.query(f'HA1_substitutions == \"{ha}\"')\n",
    "    df = df[pd.to_numeric(df['yearmonthday'], errors='coerce').notnull()]\n",
    "\n",
    "    # Identify most recent collection date\n",
    "    latest_date = (df.yearmonthday.max())\n",
    "\n",
    "    if ha == 'HA1:D7G,HA1:E50K,HA1:G53D,HA1:F79V,HA1:I140K,HA1:S156H':\n",
    "        print(ha, df)\n",
    "\n",
    "    # Select strain with most recent date\n",
    "    representative_strain_df = df.query(f'yearmonthday == {latest_date}').dropna()\n",
    "    \n",
    "    if (len(representative_strain_df)) != 1:\n",
    "        print(f'multiple strains were collected on this date, {latest_date}!')\n",
    "        \n",
    "    new_HA_strains_df = pd.concat([new_HA_strains_df, representative_strain_df])\n",
    "    \n",
    "print('here are some the representative strains for each substitution, chosen by latest collection date...')\n",
    "new_HA_strains_df = (new_HA_strains_df\n",
    "                     .reset_index(drop=True)\n",
    "                     .assign(method = 'newHAsubs')\n",
    "                    )\n",
    "new_HA_strains_df.name.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b806dc77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "writing merged metadata to ../results/auspice_and_newHAsubs_library_strains.csv...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>subclade</th>\n",
       "      <th>method</th>\n",
       "      <th>date</th>\n",
       "      <th>date_submitted</th>\n",
       "      <th>accession_ha</th>\n",
       "      <th>HA1_haplotype_count</th>\n",
       "      <th>HA1_substitutions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Finland/391/2023</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>six_month</td>\n",
       "      <td>2023-07-07</td>\n",
       "      <td>2023-09-08</td>\n",
       "      <td>EPI2736643</td>\n",
       "      <td>2708.0</td>\n",
       "      <td>HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Kanagawa/IC2239/2023</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>six_month</td>\n",
       "      <td>2023-06-27</td>\n",
       "      <td>2023-07-25</td>\n",
       "      <td>EPI2647253</td>\n",
       "      <td>11.0</td>\n",
       "      <td>HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K,HA1:K326R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A/Singapore/NUH0526/2023</td>\n",
       "      <td>H</td>\n",
       "      <td>six_month</td>\n",
       "      <td>2023-06-14</td>\n",
       "      <td>2023-07-10</td>\n",
       "      <td>EPI2617594</td>\n",
       "      <td>1626.0</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A/Sydney/332/2023</td>\n",
       "      <td>H</td>\n",
       "      <td>six_month</td>\n",
       "      <td>2023-06-14</td>\n",
       "      <td>2023-07-24</td>\n",
       "      <td>EPI2643073</td>\n",
       "      <td>274.0</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/Bangkok/P3755/2023</td>\n",
       "      <td>H</td>\n",
       "      <td>six_month</td>\n",
       "      <td>2023-09-07</td>\n",
       "      <td>2023-09-28</td>\n",
       "      <td>EPI2760070</td>\n",
       "      <td>10.0</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>A/Kanagawa/AC2316/2023</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>newHAsubs</td>\n",
       "      <td>2023-10-08</td>\n",
       "      <td>2023-11-01</td>\n",
       "      <td>EPI2781715</td>\n",
       "      <td>10.0</td>\n",
       "      <td>HA1:D104G,HA1:S124R,HA1:I140K,HA1:K276R,HA1:R299K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>A/Brisbane/429/2023</td>\n",
       "      <td>G.1.1.2</td>\n",
       "      <td>newHAsubs</td>\n",
       "      <td>2023-07-13</td>\n",
       "      <td>2023-11-09</td>\n",
       "      <td>EPI2790312</td>\n",
       "      <td>5.0</td>\n",
       "      <td>HA1:D104G,HA1:I140K,HA1:I214T,HA1:K276R,HA1:R299K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>A/SouthAfrica/R07073/2023</td>\n",
       "      <td>G.2</td>\n",
       "      <td>newHAsubs</td>\n",
       "      <td>2023-06-19</td>\n",
       "      <td>2023-09-04</td>\n",
       "      <td>EPI2725091</td>\n",
       "      <td>5.0</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:F79V,HA1:I140K,HA1:S156H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>A/SouthSudan/631/2023</td>\n",
       "      <td>G.1.3</td>\n",
       "      <td>newHAsubs</td>\n",
       "      <td>2023-06-09</td>\n",
       "      <td>2023-10-03</td>\n",
       "      <td>EPI2760859</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HA1:G53N,HA1:N96S,HA1:I192F,HA1:K276E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>A/Southafrica/R07876/2020</td>\n",
       "      <td>H</td>\n",
       "      <td>newHAsubs</td>\n",
       "      <td>2023-07-07</td>\n",
       "      <td>2023-09-06</td>\n",
       "      <td>EPI2732515</td>\n",
       "      <td>7.0</td>\n",
       "      <td>HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         name subclade     method        date date_submitted  \\\n",
       "0          A/Finland/391/2023  G.1.1.2  six_month  2023-07-07     2023-09-08   \n",
       "1      A/Kanagawa/IC2239/2023  G.1.1.2  six_month  2023-06-27     2023-07-25   \n",
       "2    A/Singapore/NUH0526/2023        H  six_month  2023-06-14     2023-07-10   \n",
       "3           A/Sydney/332/2023        H  six_month  2023-06-14     2023-07-24   \n",
       "4        A/Bangkok/P3755/2023        H  six_month  2023-09-07     2023-09-28   \n",
       "..                        ...      ...        ...         ...            ...   \n",
       "57     A/Kanagawa/AC2316/2023  G.1.1.2  newHAsubs  2023-10-08     2023-11-01   \n",
       "58        A/Brisbane/429/2023  G.1.1.2  newHAsubs  2023-07-13     2023-11-09   \n",
       "59  A/SouthAfrica/R07073/2023      G.2  newHAsubs  2023-06-19     2023-09-04   \n",
       "60      A/SouthSudan/631/2023    G.1.3  newHAsubs  2023-06-09     2023-10-03   \n",
       "61  A/Southafrica/R07876/2020        H  newHAsubs  2023-07-07     2023-09-06   \n",
       "\n",
       "   accession_ha  HA1_haplotype_count  \\\n",
       "0    EPI2736643               2708.0   \n",
       "1    EPI2647253                 11.0   \n",
       "2    EPI2617594               1626.0   \n",
       "3    EPI2643073                274.0   \n",
       "4    EPI2760070                 10.0   \n",
       "..          ...                  ...   \n",
       "57   EPI2781715                 10.0   \n",
       "58   EPI2790312                  5.0   \n",
       "59   EPI2725091                  5.0   \n",
       "60   EPI2760859                  8.0   \n",
       "61   EPI2732515                  7.0   \n",
       "\n",
       "                                    HA1_substitutions  \n",
       "0             HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K  \n",
       "1   HA1:D104G,HA1:I140K,HA1:K276R,HA1:R299K,HA1:K326R  \n",
       "2   HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...  \n",
       "3   HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...  \n",
       "4   HA1:E50K,HA1:G53N,HA1:N96S,HA1:N122D,HA1:I140K...  \n",
       "..                                                ...  \n",
       "57  HA1:D104G,HA1:S124R,HA1:I140K,HA1:K276R,HA1:R299K  \n",
       "58  HA1:D104G,HA1:I140K,HA1:I214T,HA1:K276R,HA1:R299K  \n",
       "59  HA1:E50K,HA1:G53N,HA1:F79V,HA1:I140K,HA1:S156H...  \n",
       "60              HA1:G53N,HA1:N96S,HA1:I192F,HA1:K276E  \n",
       "61  HA1:E50K,HA1:G53N,HA1:N96S,HA1:I140K,HA1:I192F...  \n",
       "\n",
       "[62 rows x 8 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine everything\n",
    "auspice_and_newHAsubs_df = (pd.concat([auspice_nextclade_merged_df, new_HA_strains_df])\n",
    "                            .reset_index(drop=True)\n",
    "                            [['name', 'subclade', 'method', 'date', 'date_submitted', \n",
    "                              'accession_ha','HA1_haplotype_count', 'HA1_substitutions']]\n",
    "                           )\n",
    "auspice_and_newHAsubs_df['accession_ha'] = auspice_and_newHAsubs_df['accession_ha'].str.replace('EPIEPI','EPI')\n",
    "\n",
    "\n",
    "outfile = os.path.join(resultsdir, 'auspice_and_newHAsubs_library_strains.csv')\n",
    "print(f'writing merged metadata to {outfile}...')\n",
    "auspice_and_newHAsubs_df.to_csv(outfile, index=False)\n",
    "auspice_and_newHAsubs_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "588acba6-7cfa-42f4-93f0-0daec55062f1",
   "metadata": {},
   "source": [
    "## Write GISIAD IDs to file\n",
    "The search function in GISAID accepts tab delimited EPI IDs. Save a file so that these IDs can more easily be copy-pasted into search box. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fcbc36ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of [Auspice/Nextclade + New HA1 subs methods] unique accessions is... 62\n",
      "here they are:  ['EPI2736643', 'EPI2647253', 'EPI2617594', 'EPI2643073', 'EPI2760070', 'EPI2735889', 'EPI2759946', 'EPI2743304', 'EPI2777879', 'EPI2689677', 'EPI2756864', 'EPI2781036', 'EPI2778022', 'EPI2735103', 'EPI2758716', 'EPI2791811', 'EPI2498763', 'EPI2505314', 'EPI2571554', 'EPI2558520', 'EPI2498814', 'EPI2312127', 'EPI2793114', 'EPI2724875', 'EPI2785896', 'EPI2771080', 'EPI2725146', 'EPI2759881', 'EPI2762684', 'EPI2773303', 'EPI2686112', 'EPI2687991', 'EPI2755805', 'EPI2688989', 'EPI2756778', 'EPI2772585', 'EPI2653147', 'EPI2724869', 'EPI2794043', 'EPI2718764', 'EPI2688666', 'EPI2736695', 'EPI2700761', 'EPI2688685', 'EPI2689110', 'EPI2775314', 'EPI2743993', 'EPI2760054', 'EPI2718582', 'EPI2670258', 'EPI2667823', 'EPI2762970', 'EPI2767381', 'EPI2763780', 'EPI2790753', 'EPI2725748', 'EPI2689865', 'EPI2781715', 'EPI2790312', 'EPI2725091', 'EPI2760859', 'EPI2732515']\n",
      "writing accession numbers to ../data/gisaid_query/library_accession_numbers.csv...\n"
     ]
    }
   ],
   "source": [
    "# Make out directory\n",
    "gisaiddir = os.path.join(datadir, 'gisaid_query')\n",
    "os.makedirs(gisaiddir, exist_ok=True)\n",
    "\n",
    "# Save just EPI IDs to CSV to make GISAID searching easy\n",
    "outfile = os.path.join(gisaiddir, 'library_accession_numbers.csv')\n",
    "library_strain_accession = auspice_and_newHAsubs_df['accession_ha']\n",
    "print(f'the number of [Auspice/Nextclade + New HA1 subs methods] unique accessions is... {len(library_strain_accession.drop_duplicates())}')\n",
    "print('here they are: ', library_strain_accession.to_list())\n",
    "print(f'writing accession numbers to {outfile}...')\n",
    "with open(outfile, 'w') as f:\n",
    "    for item in library_strain_accession:\n",
    "        f.write(item + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fda22a24",
   "metadata": {},
   "source": [
    "## Analyze HA1 mutations\n",
    "It would be potentially helpful to know how many mutatiosn are sampled in our libraries, and in what background. \n",
    "I will calculate and save that information here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2b53367a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "here are the unique mutations:  ['D104G', 'I140K', 'K276R', 'R299K', 'K326R', 'E50K', 'G53N', 'N96S', 'I192F', 'I223V', 'N122D', 'S144N', 'K276E', 'S124N', 'S145N', 'T28A', 'Q173R', 'P21T', 'I242M', 'I25V', 'I140M', 'G53D', 'F79V', 'S156H', 'G5E', 'T135A', 'S262N', 'V309I', 'E83K', 'I48T', 'R269K', 'D101E', 'D7G', 'D104N', 'A106V', 'Q327R', 'Q197H', 'I300V', 'S198P', 'K207R', 'K278E', 'G5V', 'G78D', 'N81D', 'Q173L', 'A212T', 'N63D', 'L15I', 'T10M', 'R33Q', 'G275S', 'G275D', 'T128I', 'V112I', 'K278N', 'P289L', 'K121E', 'S124R', 'I214T', 'P239S']\n",
      "writing mutations to ../results/mutation_counts_contexts.csv...\n"
     ]
    }
   ],
   "source": [
    "# Analyze HA1 mutations\n",
    "auspice_and_newHAsubs_df['HA1_muts'] = auspice_and_newHAsubs_df['HA1_substitutions'].str.replace('HA1:','')\n",
    "HA_df = auspice_and_newHAsubs_df[['HA1_muts']]\n",
    "\n",
    "# Make list of unique HA1 mutations\n",
    "unique_mutations = []\n",
    "for item in HA_df['HA1_muts'].to_list():\n",
    "    for sub in item.split(','):\n",
    "        if sub not in unique_mutations:\n",
    "            unique_mutations.append(sub)\n",
    "print('here are the unique mutations: ', unique_mutations)\n",
    "\n",
    "# Count unique contexts for each mutation\n",
    "mutation_contexts = []\n",
    "\n",
    "for mut in unique_mutations:\n",
    "    count = 0 \n",
    "    backgrounds = []\n",
    "    for item in HA_df['HA1_muts'].to_list():\n",
    "              \n",
    "        if mut in item:\n",
    "            count += 1\n",
    "            backgrounds.append(item)\n",
    "        \n",
    "    mutation_contexts.append([mut, count, backgrounds])\n",
    "    \n",
    "mutation_contexts_df = (pd.DataFrame(mutation_contexts, columns = ['mutation', 'count', 'backgrounds'])\n",
    "                        .sort_values(by=['count'], ascending=False)\n",
    "                        .reset_index(drop=True)\n",
    "                        .assign(site = lambda x: x['mutation'].str[1:-1].astype(int))\n",
    "                       )\n",
    "\n",
    "outfile = os.path.join(resultsdir, 'mutation_counts_contexts.csv')\n",
    "print(f'writing mutations to {outfile}...')\n",
    "mutation_contexts_df.to_csv(outfile, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bea1b4a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
