{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d59d6768",
   "metadata": {},
   "source": [
    "# Designing H3 HA constructs\n",
    "Here, I will design HA genes to be ordered from Twist that contain regions that overlap with \n",
    "* 2851 cut vector\n",
    "* fixed c-terminal endo domain, containing H3 and WSN recoded sequence, and a 16-nucleotide barcode\n",
    "I will then save the expected full plasmid sequence as a GenBank file. \n",
    "\n",
    "Our H3 ectodomain sequences must have overlapping regions with WSN signal peptide and packaging signal, such that we can assemble them into the vector 2851 with the endodomain/barcoded region that I will clone separately. \n",
    "\n",
    "Each ectodomain will include HA-aa-1 to HA-aa-504 with the following regions added:\n",
    "* 3' WSN signal peptide: catttgtagctacagatgcagacaca\n",
    "* 5' H3/WSN packaging signal: gttgagctgaagtcaggatac\n",
    "We will order these from Twist as gBlocks.\n",
    "\n",
    "Author: Caroline Kikawa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cb9d550a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import relevant packages\n",
    "import os\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9aa99a14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ID input and output\n",
    "datadir = '../data'\n",
    "resultsdir = '../results'\n",
    "\n",
    "os.makedirs(datadir, exist_ok=True)\n",
    "os.makedirs(resultsdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15d9c5bf",
   "metadata": {},
   "source": [
    "## Design vaccine strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "58b08c93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ID alignment files\n",
    "# Code expects these to be aligned and have the 3' NCR deleted\n",
    "vaccines_strains_file = os.path.join(datadir, 'alignments/vaccine_strains_HA_nc_ALIGN_EDIT.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2954da0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize empty list of sequence constructs\n",
    "vaccine_H3_constructs = []\n",
    "\n",
    "# Update on Jan 6, 2024\n",
    "# Initialize empty list of sequence constructs for second order\n",
    "vaccine_H3_constructs_2ndOrder = []\n",
    "\n",
    "# Define signal peptide and packaging sequences to add to each construct\n",
    "WSN_signal_peptide_overlap = 'catttgtagctacagatgcagacaca'\n",
    "H3_WSN_packaging_overlap = 'gttgagctgaagtcaggatac'\n",
    "\n",
    "# Iterate through H3 ectodomains\n",
    "# Trim 3' packaging signal and 3' transmembrane regions\n",
    "# Add WSN signal peptide and H3-WSN TM/packaging sequences\n",
    "fasta_sequences = SeqIO.parse(open(vaccines_strains_file),'fasta')\n",
    "for fasta in fasta_sequences:\n",
    "    name, sequence = fasta.id, str(fasta.seq)\n",
    "    \n",
    "    # Check that sequences start with start codon (ie., is 3' NCR trimmed?)\n",
    "    if (sequence[0:3]) != 'atg':\n",
    "        print(f'! unexpected initial codon in {name}')\n",
    "        \n",
    "    # Build construct\n",
    "    HA1 = sequence[48:1560]\n",
    "    construct = WSN_signal_peptide_overlap + HA1 + H3_WSN_packaging_overlap\n",
    "    \n",
    "    # More succinct name to put on order sheet\n",
    "    shortname = (' '.join(name.split('|')[0:2])\n",
    "          .replace('_','-')\n",
    "          .strip('--')\n",
    "          .replace('EPI-ISL-','')\n",
    "          .replace('--(18/242)','')\n",
    "          .replace('-(15/192)','')\n",
    "          .replace('--(20/322)','')\n",
    "          .replace('--(19/292)','')\n",
    "          .replace(' ','_')\n",
    "                 \n",
    "        # These names are too long and need additional shortening\n",
    "          .replace('Singapore', 'Sing')\n",
    "          .replace('-X-307A_293527', '_egg')\n",
    "          .replace('225834', 'cell')\n",
    "          .replace('Switzerland', 'Switz')\n",
    "          .replace('198223', 'egg')\n",
    "          .replace('166310', 'cell')\n",
    "          # Texas\n",
    "          .replace('-(13/252)_215972', '_egg')\n",
    "          .replace('122006', 'cell'))\n",
    "    \n",
    "        \n",
    "    # For vaccine strains, we can skip ordering Cambodia/e0826360/2020 (EPI_ISL_806547, egg) \n",
    "    # and Darwin/9/2021 (EPI_ISL_2233240, egg)\n",
    "    if '806547' in shortname:\n",
    "        vaccine_H3_constructs_2ndOrder.append([shortname, construct])\n",
    "    elif '2233240' in shortname:\n",
    "        vaccine_H3_constructs_2ndOrder.append([shortname, construct])\n",
    "    \n",
    "    # Add construct to list\n",
    "    vaccine_H3_constructs.append([shortname, construct])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2225969b",
   "metadata": {},
   "source": [
    "## Write ordersheet for vaccine strains\n",
    "This order was placed in October 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3a0e4de4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saving ordersheet to ../results/ordersheets/vaccine_constructs_ordersheet.csv...\n",
      "here is a preview of the ordersheet\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Switz/9715293/2013_cell</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaacttcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Switz/9715293/2013-NIB-88_egg</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaacttcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A/Texas/50/2012-X-223A_egg</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaacttcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A/Texas/50/2012_cell</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaacttcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/Hong-Kong/4801/2014_218060</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaattcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A/Hong-Kong/4801/2014_165554</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaattcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A/Hong-Kong/45/2019_347938</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaattcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A/Hong-Kong/2671/2019_400888</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaattcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A/Sing/INFIMH-16-0019/2016_cell</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaattcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A/Sing/INFIMH-16-0019/2016_egg</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaattcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A/Kansas/14/2017_292575</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaattcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A/Kansas/14/2017-X-327_346457</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaattcctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A/Cambodia/e0826360/2020_944639</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatccctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A/Darwin/6/2021_3534319</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A/Cambodia/e0826360/2020_806547</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatccctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A/Darwin/9/2021_2233240</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               name  \\\n",
       "0         A/Switz/9715293/2013_cell   \n",
       "1   A/Switz/9715293/2013-NIB-88_egg   \n",
       "2        A/Texas/50/2012-X-223A_egg   \n",
       "3              A/Texas/50/2012_cell   \n",
       "4      A/Hong-Kong/4801/2014_218060   \n",
       "5      A/Hong-Kong/4801/2014_165554   \n",
       "6        A/Hong-Kong/45/2019_347938   \n",
       "7      A/Hong-Kong/2671/2019_400888   \n",
       "8   A/Sing/INFIMH-16-0019/2016_cell   \n",
       "9    A/Sing/INFIMH-16-0019/2016_egg   \n",
       "10          A/Kansas/14/2017_292575   \n",
       "11    A/Kansas/14/2017-X-327_346457   \n",
       "12  A/Cambodia/e0826360/2020_944639   \n",
       "13          A/Darwin/6/2021_3534319   \n",
       "14  A/Cambodia/e0826360/2020_806547   \n",
       "15          A/Darwin/9/2021_2233240   \n",
       "\n",
       "                                             sequence  \n",
       "0   catttgtagctacagatgcagacacacaaaaacttcctggaaatga...  \n",
       "1   catttgtagctacagatgcagacacacaaaaacttcctggaaatga...  \n",
       "2   catttgtagctacagatgcagacacacaaaaacttcctggaaatga...  \n",
       "3   catttgtagctacagatgcagacacacaaaaacttcctggaaatga...  \n",
       "4   catttgtagctacagatgcagacacacaaaaaattcctggaaatga...  \n",
       "5   catttgtagctacagatgcagacacacaaaaaattcctggaaatga...  \n",
       "6   catttgtagctacagatgcagacacacaaaaaattcctggaaatga...  \n",
       "7   catttgtagctacagatgcagacacacaaaaaattcctggaaatga...  \n",
       "8   catttgtagctacagatgcagacacacaaaaaattcctggaaatga...  \n",
       "9   catttgtagctacagatgcagacacacaaaaaattcctggaaatga...  \n",
       "10  catttgtagctacagatgcagacacacaaaaaattcctggaaatga...  \n",
       "11  catttgtagctacagatgcagacacacaaaaaattcctggaaatga...  \n",
       "12  catttgtagctacagatgcagacacacaaaaaatccctggaaatga...  \n",
       "13  catttgtagctacagatgcagacacacaaaaaatacctggaaatga...  \n",
       "14  catttgtagctacagatgcagacacacaaaaaatccctggaaatga...  \n",
       "15  catttgtagctacagatgcagacacacaaaaaatacctggaaatga...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # Save fasta file of constructs\n",
    "# outfile = os.path.join(resultsdir, 'vaccine_constructs.fasta')\n",
    "# with open(outfile, 'w') as f:\n",
    "#     for item in vaccine_H3_constructs:\n",
    "#         name = item[0]\n",
    "#         seq = item[1]\n",
    "#         f.write('>' + name + '\\n')\n",
    "#         f.write(seq + '\\n')\n",
    "# print(f'saving FASTA format HA construct sequences to {outfile}...')\n",
    "\n",
    "\n",
    "# Make dataframe for ordersheet\n",
    "vaccine_H3_constructs_df = pd.DataFrame(vaccine_H3_constructs, columns = ['name', 'sequence'])\n",
    "\n",
    "# Save ordersheet to results\n",
    "outfile = os.path.join(resultsdir, 'ordersheets/vaccine_constructs_ordersheet.csv')\n",
    "vaccine_H3_constructs_df.to_csv(outfile, index=False)\n",
    "print(f'saving ordersheet to {outfile}...')\n",
    "\n",
    "# Preview df \n",
    "print('here is a preview of the ordersheet')\n",
    "vaccine_H3_constructs_df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3397126f",
   "metadata": {},
   "source": [
    "## Design library strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d5003b82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ID alignment files\n",
    "# Code expects these to be aligned and have the 3' NCR deleted\n",
    "library_strains_file = os.path.join(datadir, 'alignments/library_strains.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6fa31b3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "! unexpected initial codon in A/Ontario/RV00796/2023|EPI_ISL_18234906|A_/_H3N2|P1||2023-07-25/1-1653\n"
     ]
    }
   ],
   "source": [
    "# Initialize empty list of sequence constructs\n",
    "library_H3_constructs = []\n",
    "\n",
    "# Define signal peptide and packaging sequences to add to each construct\n",
    "WSN_signal_peptide_overlap = 'catttgtagctacagatgcagacaca'\n",
    "H3_WSN_packaging_overlap = 'gttgagctgaagtcaggatac'\n",
    "\n",
    "# Iterate through H3 ectodomains\n",
    "# Trim 3' packaging signal and 3' transmembrane regions\n",
    "# Add WSN signal peptide and H3-WSN TM/packaging sequences\n",
    "fasta_sequences = SeqIO.parse(open(library_strains_file),'fasta')\n",
    "for fasta in fasta_sequences:\n",
    "    name, sequence = fasta.id, str(fasta.seq)\n",
    "    \n",
    "    # Check that sequences start with start codon (ie., is 3' NCR trimmed?)\n",
    "    if (sequence[0:3]) != 'atg':\n",
    "        print(f'! unexpected initial codon in {name}')\n",
    "    \n",
    "    # Build construct\n",
    "    HA1 = sequence[48:1560]\n",
    "    construct = WSN_signal_peptide_overlap + HA1 + H3_WSN_packaging_overlap\n",
    "    \n",
    "    # More succinct name to put on order sheet\n",
    "    shortname = (name.split('|')[0]\n",
    "                 # These names are too long and need additional shortening\n",
    "                 .replace('Saint-Petersburg', 'Saint-Peter')\n",
    "                )\n",
    "        \n",
    "    # Add construct to list\n",
    "    library_H3_constructs.append([shortname, construct])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed0e8b5a",
   "metadata": {},
   "source": [
    "## Write ordersheet for currently circulating library strains\n",
    "This order was placed in November 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a550de58",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saving ordersheet to ../results/ordersheets/library_constructs_ordersheet.csv...\n",
      "here is a preview of the ordersheet\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Abu_Dhabi/6753/2023</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Georgia/40/2023</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A/California/81/2023</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A/Wisconsin/27/2023</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/Maldives/852/2023</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    name                                           sequence\n",
       "0  A/Abu_Dhabi/6753/2023  catttgtagctacagatgcagacacacaaaaaatacctggaaatga...\n",
       "1      A/Georgia/40/2023  catttgtagctacagatgcagacacacaaaaaatacctggaaatga...\n",
       "2   A/California/81/2023  catttgtagctacagatgcagacacacaaaaaatacctggaaatga...\n",
       "3    A/Wisconsin/27/2023  catttgtagctacagatgcagacacacaaaaaatacctggaaatga...\n",
       "4    A/Maldives/852/2023  catttgtagctacagatgcagacacacaaaaaatacctggaaatga..."
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # Save fasta file of constructs\n",
    "# outfile = os.path.join(resultsdir, 'library_constructs.fasta')\n",
    "# with open(outfile, 'w') as f:\n",
    "#     for item in library_H3_constructs:\n",
    "#         name = item[0]\n",
    "#         seq = item[1]\n",
    "#         f.write('>' + name + '\\n')\n",
    "#         f.write(seq + '\\n')\n",
    "# print(f'saving FASTA format HA construct sequences to {outfile}...')\n",
    "\n",
    "# Make dataframe for ordersheet\n",
    "library_H3_constructs_df = pd.DataFrame(library_H3_constructs, columns = ['name', 'sequence'])\n",
    "\n",
    "# Save ordersheet to results\n",
    "outfile = os.path.join(resultsdir, 'ordersheets/library_constructs_ordersheet.csv')\n",
    "library_H3_constructs_df.to_csv(outfile, index=False)\n",
    "print(f'saving ordersheet to {outfile}...')\n",
    "\n",
    "# Preview df \n",
    "print('here is a preview of the ordersheet')\n",
    "library_H3_constructs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aafb0450",
   "metadata": {},
   "source": [
    "## Add most recent 2024 vaccine strains\n",
    "I need to design constructs for the most recent 2024 vaccine strains. I'll add these to the 2nd order sheet list initialized for the Darwin/9 and Cambodia strains above. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cb884743",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_vaccine_strains_file = os.path.join(datadir, 'alignments/2024_vacc_HA_nc_ALIGN.fasta')\n",
    "\n",
    "# Iterate through H3 ectodomains\n",
    "# Trim 3' packaging signal and 3' transmembrane regions\n",
    "# Add WSN signal peptide and H3-WSN TM/packaging sequences\n",
    "fasta_sequences = SeqIO.parse(open(new_vaccine_strains_file),'fasta')\n",
    "for fasta in fasta_sequences:\n",
    "    name, sequence = fasta.id, str(fasta.seq)\n",
    "    \n",
    "    # Check that sequences start with start codon (ie., is 3' NCR trimmed?)\n",
    "    if (sequence[0:3]) != 'atg':\n",
    "        print(f'! unexpected initial codon in {name}')\n",
    "        \n",
    "    # Build construct\n",
    "    HA1 = sequence[48:1560]\n",
    "    construct = WSN_signal_peptide_overlap + HA1 + H3_WSN_packaging_overlap\n",
    "    \n",
    "    # More succinct name to put on order sheet\n",
    "    shortname = shortname = (name.split('|')[0])\n",
    "    \n",
    "    # Add construct to new order list\n",
    "    vaccine_H3_constructs_2ndOrder.append([shortname, construct])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47615358",
   "metadata": {},
   "source": [
    "## Write ordersheet for new vaccine strains\n",
    "This order was placed in January 2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3618e6b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saving ordersheet to ../results/ordersheets/2024VaccineStrains_2022-2023Egg-basedVaccineStrains_ordersheet.csv...\n",
      "here is a preview of the ordersheet\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Cambodia/e0826360/2020_806547</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatccctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Darwin/9/2021_2233240</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A/Thailand/8/2022</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              name  \\\n",
       "0  A/Cambodia/e0826360/2020_806547   \n",
       "1          A/Darwin/9/2021_2233240   \n",
       "2                A/Thailand/8/2022   \n",
       "\n",
       "                                            sequence  \n",
       "0  catttgtagctacagatgcagacacacaaaaaatccctggaaatga...  \n",
       "1  catttgtagctacagatgcagacacacaaaaaatacctggaaatga...  \n",
       "2  catttgtagctacagatgcagacacacaaaaaatacctggaaatga...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # Save fasta file of constructs\n",
    "# outfile = os.path.join(resultsdir, '2024VaccineStrains_2022-2023Egg-basedVaccineStrains.fasta')\n",
    "# with open(outfile, 'w') as f:\n",
    "#     for item in vaccine_H3_constructs_2ndOrder:\n",
    "#         name = item[0]\n",
    "#         seq = item[1]\n",
    "#         f.write('>' + name + '\\n')\n",
    "#         f.write(seq + '\\n')\n",
    "# print(f'saving FASTA format HA construct sequences to {outfile}...')\n",
    "\n",
    "# Make dataframe for ordersheet\n",
    "vaccine_H3_constructs_2ndOrder_df = pd.DataFrame(vaccine_H3_constructs_2ndOrder, columns = ['name', 'sequence'])\n",
    "\n",
    "# Save ordersheet to results\n",
    "outfile = os.path.join(resultsdir, 'ordersheets/2024VaccineStrains_2022-2023Egg-basedVaccineStrains_ordersheet.csv')\n",
    "vaccine_H3_constructs_2ndOrder_df.to_csv(outfile, index=False)\n",
    "print(f'saving ordersheet to {outfile}...')\n",
    "\n",
    "# Preview df \n",
    "print('here is a preview of the ordersheet')\n",
    "vaccine_H3_constructs_2ndOrder_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44462273",
   "metadata": {},
   "source": [
    "## Saving expected sequence as FASTA files\n",
    "I can use these files as input in Benchling for sequence confirmation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8b82f791",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "! unexpected initial codon in A/Ontario/RV00796/2023|EPI_ISL_18234906|A_/_H3N2|P1||2023-07-25/1-1653\n",
      "saving construct protein sequences to ../results/2023-2024_H3_library_protein_constructs.fasta...\n"
     ]
    }
   ],
   "source": [
    "# mapsdir = os.path.join(resultsdir, 'construct_maps')\n",
    "# os.makedirs(mapsdir, exist_ok=True)\n",
    "\n",
    "# save sequence upstream and downstream (including expected overlap)\n",
    "upstream_vector = 'actcttcctttttcaatattattgaagcatttatcagggttattgtctcatgagcggatacatatttgaatgtatttagaaaaataaacaaaagagtttgtagaaacgcaaaaaggccatccgtcaggatggccttctgcttaatttgatgcctggcagtttatggcgggcgtcctgcccgccaccctccgggccgttgcttcgcaacgttcaaatccgctcccggcggatttgtcctactcaggagagcgttcaccgacaaacaacagataaaacgaaaggcccagtctttcgactgagcctttcgttttatttgatgcctggcagttccctactctcgcatggggagaccccacactaccatcggcgctacggcgtttcacttctgagttcggcatggggtcaggtgggaccaccgcgctactgccgccaggcaaattctgttttatcagaccgcttctgcgttctgatttaatctgtatcaggctgaaaattttttttcatccgccaaaacagccaaggcggccgcgctagcggccgatccccaaaaaaaaaaaaaaaaaaagagtccagagtggccccgccgctccgcgccggggggggggggggggggggacactttcggacatctggtcgacctccagcatcgggggaaaaaaaaaaacaaagtgtcgcccggagtactggtcgacctccgaagttgggggggagcaaaagcaggggaaaataaaaacaaccaaaatgaaggcaaaactactggtcctgttatatgcatttgtagctacagatgcagacaca'\n",
    "endodomain = 'gttgagctgaagtcaggatacaaagattggatcctatggatttcctttgccATGtcTtgCttCCtActGtgCgtAgcACtACtAggCttTatTatgtgggcGtgTcaGaaAggCtcCCtAcaAtgTCgGatTtgTatTTAATAG'\n",
    "barcode = 'NNNNNNNNNNNNNNNN'\n",
    "downstream_vector = 'agatcggaagagcgtcgtgtagggaaagagtgtgcggccgctatctactcaactgtcgccagttcactggtgctttaggtctccctgggggcaatcagtttctggatgtgttctaatgggtctttgcagtgcagaatatgcatctgagattaggatttcagaaatataaggaaaaacacccttgtttctactaataacccggcggcccaaaatgccgactcggagcgaaagatatacctcccccggggccgggaggtcgcgtcaccgaccacgccgccggcccaggcgacgcgcgacacggacacctgtccccaaaaacgccaccatcgcagccacacacggagcgcccggggccctctggtcaaccccaggacacacgcgggagcagcgccgggccggggacgccctcccggccgcccgtgccacacgcagggggccggcccgtgtctccagagcgggagccggaagcattttcggccggcccctcctacgaccgggacacacgagggaccgaaggccggccaggcgcgacctctcgggccgcacgcgcgctcagggagcgctctccgactccgcacggggactcgccagaaaggatcgtgatctgcattaatgaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaagaacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgagattatcaaaaaggatcttcacctagatccttttaaattaaaaatgaagttttaaatcaatctaaagtatatatgagtaaacttggtctgacagttaccaatgcttaatcagtgaggcacctatctcagcgatctgtctatttcgttcatccatagttgcctgactccccgtcgtgtagataactacgatacgggagggcttaccatctggccccagtgctgcaatgataccgcgagacccacgctcaccggctccagatttatcagcaataaaccagccagccggaagggccgagcgcagaagtggtcctgcaactttatccgcctccatccagtctattaattgttgccgggaagctagagtaagtagttcgccagttaatagtttgcgcaacgttgttgccattgctacaggcatcgtggtgtcacgctcgtcgtttggtatggcttcattcagctccggttcccaacgatcaaggcgagttacatgatcccccatgttgtgcaaaaaagcggttagctccttcggtcctccgatcgttgtcagaagtaagttggccgcagtgttatcactcatggttatggcagcactgcataattctcttactgtcatgccatccgtaagatgcttttctgtgactggtgagtactcaaccaagtcattctgagaatagtgtatgcggcgaccgagttgctcttgcccggcgtcaacacgggataataccgcgccacatagcagaactttaaaagtgctcatcattggaaaacgttcttcggggcgaaaactctcaaggatcttaccgctgttgagatccagttcgatgtaacccactcgtgcacccaactgatcttcagcatcttttactttcaccagcgtttctgggtgagcaaaaacaggaaggcaaaatgccgcaaaaaagggaataagggcgacacggaaatgttgaatactcat'\n",
    "\n",
    "# Initialize numbering scheme in library\n",
    "libID = 1\n",
    "\n",
    "# Initialize plasmid log numbering \n",
    "logID = 4142\n",
    "lib_ID_list = []\n",
    "\n",
    "# initialize list for protein constructs\n",
    "library_H3_protein_constructs = []\n",
    "\n",
    "# Do this for both vaccine and library files\n",
    "files = [vaccines_strains_file, library_strains_file, new_vaccine_strains_file]\n",
    "\n",
    "for f in files:\n",
    "    fasta_sequences = SeqIO.parse(open(f),'fasta')\n",
    "    for fasta in fasta_sequences:\n",
    "        name, sequence = fasta.id, str(fasta.seq)\n",
    "\n",
    "        # Check that sequences start with start codon (ie., is 3' NCR trimmed?)\n",
    "        if (sequence[0:3]) != 'atg':\n",
    "            print(f'! unexpected initial codon in {name}')\n",
    "            \n",
    "        # Build construct\n",
    "        HA1 = sequence[48:1560]\n",
    "        construct_seq = (upstream_vector + \n",
    "                         HA1 + \n",
    "                         endodomain + \n",
    "                         barcode + \n",
    "                         downstream_vector)\n",
    "        \n",
    "        # Information for library \n",
    "        strain_name = name.split('|')[0]\n",
    "        epi = name.split('|')[1]\n",
    "        \n",
    "        # Rename egg-passaged Cambodia strain so it's not the exact match to the cell-passaged strain\n",
    "        if epi == 'EPI_ISL_806547':\n",
    "            strain_name = 'A/Cambodia/e0826360/2020_egg'\n",
    "\n",
    "        # Rename A/Singapore/NUH0526/2023 to A/Massachusetts/18/2022\n",
    "        # Exact amino acid match \n",
    "        if strain_name == 'A/Singapore/NUH0526/2023':\n",
    "            strain_name = 'A/Massachusetts/18/2022'\n",
    "        \n",
    "        # Need to reindex log ID for A/Thailand/8/2022, which was ordered \n",
    "        # after all other strains in library\n",
    "        if epi == 'EPI_ISL_16014504':\n",
    "            logID = 4636\n",
    "\n",
    "        # Correct GISAID- or Nextclade-generated typos (...) in a few names\n",
    "        # See issue #5, #6 and #7 in my GitHub repo\n",
    "        # https://github.com/jbloomlab/flu_seqneut_H3N2_2023-2024/issues\n",
    "        # if strain_name == 'A/SOUTH_AFRICA/R07876/202023':\n",
    "        #     strain_name = 'A/SOUTH_AFRICA/R07876/2023'\n",
    "\n",
    "        if strain_name == 'A/South_Africa/KO56863/2023':\n",
    "            strain_name = 'A/South_Africa/K056863/2023'\n",
    "\n",
    "        if strain_name == 'A/Saint-Petersburg/RII-MH144113/2023':\n",
    "            strain_name = 'A/Saint-Petersburg/RII-166/2023'\n",
    "    \n",
    "        # Replace strain names for a few vaccine strains so they match the TimeTree strain names\n",
    "        # These are amino acid identical strains, but with more sensical collection dates\n",
    "        # The original downloaded sequences had collection dates in 2016, which messed up my time tree\n",
    "        if strain_name == 'A/Texas/50/2012_X-223A_(13/252)':\n",
    "            strain_name = 'A/Texas/50/2012X-223A'\n",
    "\n",
    "        if strain_name == 'A/Hong_Kong/4801/2014_(15/192)':\n",
    "            strain_name = 'A/HongKong/4801/2014egg'\n",
    "\n",
    "            \n",
    "        # Make a name with underscores\n",
    "        underscore_name = ((strain_name)\n",
    "                     .replace('/','_')\n",
    "                     .replace('(', '')\n",
    "                     .replace(')', '')\n",
    "                    )\n",
    "        \n",
    "        # Each strain has 3 barcodes\n",
    "        for n in list(range(1,4)):\n",
    "            \n",
    "            # Write plasmid name\n",
    "            plasmid_name = (str(logID) +\n",
    "                            '_phh_' +\n",
    "                            underscore_name.replace('_','-') + \n",
    "                            '_HA_WSNflank_bc' +\n",
    "                            str(n))\n",
    "            # Save to list\n",
    "            lib_ID_list.append([libID, plasmid_name, strain_name, epi, construct_seq])\n",
    "            # Add to plasmid log ID counter\n",
    "            logID += 1\n",
    "                \n",
    "\n",
    "        # Write each map to separate output\n",
    "        # fasta_name = str(libID) + '_' + underscore_name + '_H3HAlibraryConstruct'\n",
    "        # identifier_line = \">\" + strain_name + \"\\n\"\n",
    "        # output_path = os.path.join(mapsdir, f'{fasta_name}.fasta')\n",
    "        # output_file = open(output_path,'w')\n",
    "        # output_file.write(identifier_line)\n",
    "        # sequence_line = construct_seq + \"\\n\"\n",
    "        # output_file.write(sequence_line)\n",
    "        # print(f'saving {strain_name} full plasmid sequence to {output_path}...\\n')\n",
    "        # output_file.close()\n",
    "\n",
    "        # Add to counter\n",
    "        libID += 1\n",
    "        \n",
    "        # ID construct protein sequences\n",
    "        nuc_sequence = (construct_seq[741:2310+len(endodomain)])\n",
    "        prot_sequence = str(Bio.Seq.Seq(nuc_sequence).translate())\n",
    "        library_H3_protein_constructs.append([underscore_name, prot_sequence, strain_name])\n",
    "\n",
    "# Save construct protein sequences\n",
    "outfile = os.path.join(resultsdir, '2023-2024_H3_library_protein_constructs.fasta')\n",
    "print(f'saving construct protein sequences to {outfile}...')  \n",
    "with open(outfile, 'w') as f:\n",
    "    for item in library_H3_protein_constructs:\n",
    "        name = item[2].replace('_','') # Strain names from Nextstrain have all underscores stripped\n",
    "        seq = item[1]\n",
    "\n",
    "        # Ignore the egg-passaged Kansas strain\n",
    "        # This strain ended up not being included in the library\n",
    "        if name != 'A_Kansas_14_2017_X-327': \n",
    "            f.write('>' + name + '\\n')\n",
    "            f.write(seq + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c78f65e-5a3c-4ab7-a284-2d06640597f0",
   "metadata": {},
   "source": [
    "## Save HA1 and ectodomain sequences \n",
    "Create these files by doing simple trims on the constructs list we've already generated. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "480300fa-79d6-4199-aa5f-ce671a3b6b29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saving construct protein sequences to ../results/2023-2024_H3_library_protein_HA1.fasta...\n",
      "saving construct protein sequences to ../results/2023-2024_H3_library_protein_HA_ectodomain.fasta...\n"
     ]
    }
   ],
   "source": [
    "# Save HA1 protein sequences\n",
    "outfile = os.path.join(resultsdir, '2023-2024_H3_library_protein_HA1.fasta')\n",
    "print(f'saving construct protein sequences to {outfile}...')  \n",
    "with open(outfile, 'w') as f:\n",
    "    for item in library_H3_protein_constructs:\n",
    "        name = item[2].replace('_','') # Strain names from Nextstrain have all underscores stripped\n",
    "        seq = item[1][19:348]\n",
    "\n",
    "        # Ignore the egg-passaged Kansas strain\n",
    "        # This strain ended up not being included in the library\n",
    "        if name != 'A_Kansas_14_2017_X-327': \n",
    "            f.write('>' + name + '\\n')\n",
    "            f.write(seq + '\\n')\n",
    "\n",
    "# Save ectodomain protein sequences\n",
    "outfile = os.path.join(resultsdir, '2023-2024_H3_library_protein_HA_ectodomain.fasta')\n",
    "print(f'saving construct protein sequences to {outfile}...')  \n",
    "with open(outfile, 'w') as f:\n",
    "    for item in library_H3_protein_constructs:\n",
    "        name = item[2].replace('_','') # Strain names from Nextstrain have all underscores stripped\n",
    "        seq = item[1][19:520]\n",
    "\n",
    "        # Ignore the egg-passaged Kansas strain\n",
    "        # This strain ended up not being included in the library\n",
    "        if name != 'A_Kansas_14_2017_X-327': \n",
    "            f.write('>' + name + '\\n')\n",
    "            f.write(seq + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4ecc6bb-5010-463a-90f9-02956e6e2ca9",
   "metadata": {},
   "source": [
    "## Write library ID sheet\n",
    "This should have the library ID (numbers 1-79), the Bloom lab plasmid ID, the GISAID EPI and the expected construct sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "11c34ee3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saving library info sheet to ../results/library_ID_sheet.csv...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['A/Switzerland/9715293/2013', 'A/Switzerland/9715293/2013_NIB-88',\n",
       "       'A/Texas/50/2012X-223A', 'A/Texas/50/2012',\n",
       "       'A/HongKong/4801/2014egg', 'A/Hong_Kong/4801/2014',\n",
       "       'A/Hong_Kong/45/2019', 'A/Hong_Kong/2671/2019',\n",
       "       'A/Singapore/INFIMH-16-0019/2016',\n",
       "       'A/Singapore/INFIMH-16-0019/2016_X-307A', 'A/Kansas/14/2017',\n",
       "       'A/Kansas/14/2017_X-327', 'A/Cambodia/e0826360/2020',\n",
       "       'A/Darwin/6/2021', 'A/Cambodia/e0826360/2020_egg',\n",
       "       'A/Darwin/9/2021', 'A/Abu_Dhabi/6753/2023', 'A/Georgia/40/2023',\n",
       "       'A/California/81/2023', 'A/Wisconsin/27/2023',\n",
       "       'A/Maldives/852/2023', 'A/Oman/3011/2023',\n",
       "       'A/SOUTH_AFRICA/R07876/202023', 'A/Bhutan/0845/2023',\n",
       "       'A/Netherlands/01693/2023', 'A/Bangkok/P3755/2023',\n",
       "       'A/Krabi/THIS050/2023', 'A/Netherlands/01685/2023',\n",
       "       'A/Netherlands/01760/2023', 'A/Chipata/15-NIC-001/2023',\n",
       "       'A/Malaysia/IMR-SARI1989/2023', 'A/Massachusetts/18/2022',\n",
       "       'A/Victoria/1033/2023', 'A/Catalonia/NSVH102124476/2023',\n",
       "       'A/Sydney/749/2023', 'A/Townsville/68/2023', 'A/Sydney/710/2023',\n",
       "       'A/Sydney/332/2023', 'A/South_Africa/R05510/2023',\n",
       "       'A/SouthAfrica/PET28931/2023', 'A/South_Africa/R07188/2023',\n",
       "       'A/South_Africa/R06477/2023', 'A/South_Africa/R06240/2023',\n",
       "       'A/Solwezi/13-NIC-001/2023', 'A/Sydney/715/2023',\n",
       "       'A/Ontario/RV00796/2023', 'A/Samut_Songkhram/THIS061/2023',\n",
       "       'A/Bangkok/P3599/2023', 'A/Guangdong-Futian/1980/2023',\n",
       "       'A/South_Africa/K056872/2023', 'A/South_Africa/K056863/2023',\n",
       "       'A/Saskatchewan/SKFLU317847/2023',\n",
       "       'A/Saint-Petersburg/RII-166/2023', 'A/Finland/399/2023',\n",
       "       'A/SouthSudan/642/2023', 'A/SouthSudan/631/2023',\n",
       "       'A/Catalonia/2041146NS/2023', 'A/Bhutan/0006/2023',\n",
       "       'A/Bhutan/FLU-BTG-00988/2022', 'A/Finland/391/2023',\n",
       "       'A/KANAGAWA/IC2239/2023', 'A/South_Africa/R06359/2023',\n",
       "       'A/Brisbane/429/2023', 'A/KANAGAWA/AC2316/2023',\n",
       "       'A/Busan/1301/2023', 'A/TECPAN/017FLU/2023', 'A/EHIME/50/2023',\n",
       "       'A/Romania/543634/2022', 'A/YAMAGATA/98/2023',\n",
       "       'A/Schleswig-Holstein/4/2023', 'A/SENDAI/45/2023',\n",
       "       'A/South_Africa/R06506/2023', 'A/South_Africa/R05384/2023',\n",
       "       'A/South_Africa/R07073/2023', 'A/Luga/RII-11393S/2023',\n",
       "       'A/Surat_Thani/P152/2023', 'A/Jeju/1047/2023',\n",
       "       'A/South_Dakota/22/2023', 'A/Thailand/8/2022'], dtype=object)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outfile = os.path.join(resultsdir, 'library_ID_sheet.csv')\n",
    "library_ID_df = pd.DataFrame(lib_ID_list, columns = ['lib-id', 'plasmid', 'strain-name', 'epi', 'expected-seq'])\n",
    "library_ID_df.to_csv(outfile, index=False)\n",
    "print(f'saving library info sheet to {outfile}...')\n",
    "library_ID_df['strain-name'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aafb6fea",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65ff6200",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1702f3a7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
